{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9299101f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "99c8859e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('bhatt_results_nextstrain/concatenated_results_df/results_for_altair.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b71f9552",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make color mapping lists\n",
    "dom = list(set(df['legible_name']))\n",
    "rng = []\n",
    "for d in dom:\n",
    "    rng.append(list(df[df['legible_name']==d]['color'])[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e941287c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#order to plot on the x-axis\n",
    "virus_order = ['Measles', 'Mumps', 'Parainfluenza-1', 'Parainfluenza-3', 'Influenza A/H3N2', \n",
    "                  'Influenza A/H1N1pdm', 'Influenza B/Vic', 'Influenza B/Yam', \n",
    "                  'OC43-A', '229E', 'NL63', 'RSV-A', 'RSV-B', \n",
    "                  'Dengue 1-I', 'Dengue 1-III', 'Dengue 1-IV', 'Dengue 1-V', \n",
    "                  'Dengue 2-AA', 'Dengue 2-AI', 'Dengue 2-C', \n",
    "               'Dengue 3-II', 'Dengue 3-III', 'Dengue 4-I', \n",
    "                  'Dengue 4-II', 'Enterovirus D68', 'Norovirus GII.4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "527adab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_virus_config(virus):\n",
    "    config_json = f'config/adaptive_evo_config_{virus}.json'\n",
    "    with open(config_json) as json_handle:\n",
    "        configs = json.load(json_handle)\n",
    "        \n",
    "    return configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "135f25cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get name of each viral gene and make it legible\n",
    "virus_gene_names = []\n",
    "\n",
    "#add legible gene names to df\n",
    "legible_genes = {'polymerase': 'Polymerase', 'membrane_fusion': 'Membrane Fusion', \n",
    "                 'receptor_binding':'Receptor Binding'}\n",
    "\n",
    "all_viruses = list(set(df['virus']))\n",
    "for v in all_viruses:\n",
    "    configs = readin_virus_config(v)\n",
    "    for x in ['polymerase', 'membrane_fusion', 'receptor_binding']:\n",
    "        gene_name = configs[x]['virus_gene']\n",
    "        legible_gene =legible_genes[x]\n",
    "        gene_name_legible = f'{gene_name.upper()} ({legible_gene})'\n",
    "        virus_gene_names.append({'virus':v, 'gene': x, 'legible_gene_name': gene_name_legible})\n",
    "        \n",
    "virus_gene_name_mapper = pd.DataFrame(virus_gene_names)\n",
    "\n",
    "df = df.merge(virus_gene_name_mapper.reset_index())\n",
    "\n",
    "#capitalize virus family\n",
    "df['virus_family'] = df['virus_family'].str.capitalize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a1508365",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add information about genome type\n",
    "virus_genome_types = []\n",
    "for v in all_viruses:\n",
    "    configs = readin_virus_config(v)\n",
    "    genome_type = configs['genome_type']\n",
    "    enveloped = configs['enveloped']\n",
    "    virus_genome_types.append({'virus':v, 'genome_type': genome_type, 'enveloped': enveloped})\n",
    "    \n",
    "genome_type_mapper = pd.DataFrame(virus_genome_types)\n",
    "\n",
    "df = df.merge(genome_type_mapper, on= 'virus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bce6426f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add a column for the url address\n",
    "map_to_url = {'Measles': 'measles', 'Mumps':'mumps', \n",
    "              'Parainfluenza-1':'parainfluenza_1', 'Parainfluenza-3':'parainfluenza_3',\n",
    "              'Influenza A/H3N2':'h3n2', \n",
    "                  'Influenza A/H1N1pdm': 'h1n1pdm', 'Influenza B/Vic': 'vic', 'Influenza B/Yam': 'yam', \n",
    "                  'OC43-A': 'oc43_a', 'OC43-B': 'oc43_b', '229E': '229e', 'NL63':'nl63', \n",
    "              'Lassa-A':'lassa_a', 'Lassa-B':'lassa_b', 'RSV-A':'rsv_a', 'RSV-B':'rsv_b', \n",
    "                  'Dengue 1-I':'denv1_i', 'Dengue 1-III':'denv1_iii', 'Dengue 1-IV':'denv1_iv', 'Dengue 1-V':'denv1_v', \n",
    "                  'Dengue 2-AA':'denv2_aa', 'Dengue 2-AI':'denv2_ai', 'Dengue 2-AM':'denv1_am', 'Dengue 2-C':'denv2_c', \n",
    "                  'Dengue 3-I':'denv3_i', 'Dengue 3-II':'denv3_ii', 'Dengue 3-III':'denv3_iii', 'Dengue 4-I':'denv4_i', \n",
    "                  'Dengue 4-II':'denv4_ii', 'Zika':'zika', 'Enterovirus D68': 'enterovirusd68', \n",
    "                'Norovirus GII.4': 'norovirus_gii4'}\n",
    "\n",
    "df['url_ending'] = df['legible_name'].map(map_to_url)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "65e28c0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make x-axis values (to allow squishing viruses closer together or spreading them apart)\n",
    "x_tick_pos = []\n",
    "\n",
    "all_xticks = []\n",
    "\n",
    "last_coord = 0\n",
    "\n",
    "regular_spacing = 3\n",
    "squished_spacing = 1\n",
    "\n",
    "#separate enveloped and non-enveloped viruses by a bit more\n",
    "#keep track of where this is to add a vertical line\n",
    "\n",
    "#first non-enveloped virus\n",
    "first_nonenveloped = 'Enterovirus D68'\n",
    "\n",
    "\n",
    "dengue_serotypes_already_seen = []\n",
    "for x in virus_order:\n",
    "    #regular spacing between dengue serotypes, but squished spacing between genotypes of same serotype\n",
    "    if 'Dengue' in x:\n",
    "        if x[0:8] in dengue_serotypes_already_seen:\n",
    "            tick_pos = last_coord+squished_spacing\n",
    "            last_coord = tick_pos\n",
    "        else:\n",
    "            tick_pos = last_coord+regular_spacing\n",
    "            last_coord = tick_pos\n",
    "            dengue_serotypes_already_seen.append(x[0:8])\n",
    "    elif x == first_nonenveloped:\n",
    "        enveloped_vert_divider = last_coord+regular_spacing\n",
    "        tick_pos = last_coord+regular_spacing*2\n",
    "    else:\n",
    "        tick_pos = last_coord+regular_spacing\n",
    "        \n",
    "    last_coord = tick_pos\n",
    "        \n",
    "    all_xticks.append(tick_pos)\n",
    "    x_tick_pos.append({'legible_name': x, 'xtick_pos': tick_pos})\n",
    "            \n",
    "\n",
    "\n",
    "xtick_df = pd.DataFrame(x_tick_pos)\n",
    "\n",
    "df = df.merge(xtick_df, on='legible_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "490f317b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make labelExpr\n",
    "label_expr_str = \"\"\n",
    "\n",
    "for i in range(len(x_tick_pos)):\n",
    "    x = x_tick_pos[i]\n",
    "    if i==len(x_tick_pos)-1:\n",
    "        label_expr_str+=f\"'{x['legible_name']}'\"\n",
    "    else:\n",
    "        label_expr_str+=f\"datum.label == {x['xtick_pos']} ? '{x['legible_name']}' : \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5303546d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add column for legible rate\n",
    "df['legible_rate_percodon'] = df['adaptive_subs_per_codon_per_year'].round(2).astype(str)+'×10⁻³ muts per codon per year'\n",
    "df['legible_rate_pergene'] = df['adaptive_muts_per_year'].round(2).astype(str)+' muts per year'\n",
    "\n",
    "\n",
    "#subset dataframes to each gene\n",
    "df_polymerase = df[df['gene']=='polymerase']\n",
    "df_fusion = df[df['gene']=='membrane_fusion']\n",
    "df_receptorbinding = df[df['gene']=='receptor_binding']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "66389138",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get relative sizes of plot for each gene in order to make y-axis scale the same, \n",
    "#but truncate axes of polymerase and fusion\n",
    "\n",
    "p_floor = min(df_polymerase['lower_95ci'])\n",
    "p_ceiling = max(df_polymerase['upper_95ci'])\n",
    "\n",
    "f_floor = min(df_fusion['lower_95ci'])\n",
    "f_ceiling = max(df_fusion['upper_95ci'])\n",
    "\n",
    "r_floor = min(df_receptorbinding['lower_95ci'])\n",
    "r_ceiling = max(df_receptorbinding['upper_95ci'])\n",
    "\n",
    "#relative extent of y-axes (compared to receptor-binding, which has largest range)\n",
    "r_range = r_ceiling-r_floor\n",
    "p_relative_range = (p_ceiling-p_floor)/r_range\n",
    "f_relative_range = (f_ceiling-f_floor)/r_range\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dd2a8d02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-bdb4d79c1c084590afc15af93f1626ee\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-bdb4d79c1c084590afc15af93f1626ee\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-bdb4d79c1c084590afc15af93f1626ee\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300, \"strokeWidth\": 0}, \"axis\": {\"grid\": false}, \"axisX\": {\"labelAngle\": 270}, \"title\": {\"anchor\": \"start\", \"color\": \"#c6c6c6\", \"fontSize\": 18}}, \"hconcat\": [{\"data\": {\"name\": \"empty\"}, \"mark\": {\"type\": \"text\", \"align\": \"center\", \"angle\": 270, \"baseline\": \"bottom\", \"color\": \"black\", \"fontSize\": 16, \"fontWeight\": 400}, \"encoding\": {\"text\": {\"value\": [\"Rate of Adaptation\"]}, \"x\": {\"value\": 10}, \"y\": {\"value\": 160}}}, {\"data\": {\"name\": \"empty\"}, \"mark\": {\"type\": \"text\", \"align\": \"center\", \"angle\": 270, \"baseline\": \"bottom\", \"color\": \"black\", \"fontSize\": 12, \"fontWeight\": 400}, \"encoding\": {\"text\": {\"value\": [\"Adaptive Mutations per Codon per Year (\\u00d7 10\\u207b\\u00b3)\"]}, \"x\": {\"value\": 0}, \"y\": {\"value\": 160}}}, {\"vconcat\": [{\"layer\": [{\"data\": {\"name\": \"data-06a554f865048b667d6c6c7db9d5e2db\"}, \"mark\": {\"type\": \"point\", \"filled\": true, \"opacity\": 1, \"size\": 200}, \"encoding\": {\"color\": {\"field\": \"legible_name\", \"legend\": null, \"scale\": {\"domain\": [\"Influenza B/Vic\", \"229E\", \"Dengue 2-C\", \"Dengue 1-IV\", \"Influenza A/H3N2\", \"Lassa-A\", \"RSV-B\", \"RSV-A\", \"Dengue 3-II\", \"Dengue 4-I\", \"OC43-A\", \"Dengue 4-II\", \"Enterovirus D68\", \"Dengue 1-V\", \"Dengue 3-III\", \"Dengue 2-AI\", \"Influenza A/H1N1pdm\", \"Dengue 2-AA\", \"Mumps\", \"Lassa-B\", \"Parainfluenza-1\", \"Norovirus GII.4\", \"Parainfluenza-3\", \"NL63\", \"Influenza B/Yam\", \"Measles\", \"Dengue 1-III\", \"Dengue 1-I\"], \"range\": [\"#E06E85\", \"#76C7BE\", \"#2cccf2\", \"#0db0d6\", \"#9F2A44\", \"#9cc479\", \"#6C7593\", \"#0B194C\", \"#76def7\", \"#adebfa\", \"#208288\", \"#adebfa\", \"#a373a5\", \"#0db0d6\", \"#76def7\", \"#2cccf2\", \"#CC415A\", \"#2cccf2\", \"#eb520d\", \"#c4dcaf\", \"#f58451\", \"#d0b9d2\", \"#f79e76\", \"#BBE4D1\", \"#ED9AB0\", \"#F5AD52\", \"#0db0d6\", \"#0db0d6\"]}, \"type\": \"nominal\"}, \"href\": {\"field\": \"url\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"legible_name\", \"title\": \"Virus\", \"type\": \"nominal\"}, {\"field\": \"virus_family\", \"title\": \"Family\", \"type\": \"nominal\"}, {\"field\": \"genome_type\", \"title\": \"Genome Type\", \"type\": \"nominal\"}, {\"field\": \"enveloped\", \"title\": \"Enveloped\", \"type\": \"nominal\"}, {\"field\": \"legible_gene_name\", \"title\": \"Protein/Subunit\", \"type\": \"nominal\"}, {\"field\": \"legible_rate_percodon\", \"title\": \"Rate\", \"type\": \"nominal\"}], \"x\": {\"axis\": {\"labelExpr\": \"datum.label == 3 ? 'Measles' : datum.label == 6 ? 'Mumps' : datum.label == 9 ? 'Parainfluenza-1' : datum.label == 12 ? 'Parainfluenza-3' : datum.label == 15 ? 'Influenza A/H3N2' : datum.label == 18 ? 'Influenza A/H1N1pdm' : datum.label == 21 ? 'Influenza B/Vic' : datum.label == 24 ? 'Influenza B/Yam' : datum.label == 27 ? 'OC43-A' : datum.label == 30 ? '229E' : datum.label == 33 ? 'NL63' : datum.label == 36 ? 'RSV-A' : datum.label == 39 ? 'RSV-B' : datum.label == 42 ? 'Dengue 1-I' : datum.label == 43 ? 'Dengue 1-III' : datum.label == 44 ? 'Dengue 1-IV' : datum.label == 45 ? 'Dengue 1-V' : datum.label == 48 ? 'Dengue 2-AA' : datum.label == 49 ? 'Dengue 2-AI' : datum.label == 50 ? 'Dengue 2-C' : datum.label == 53 ? 'Dengue 3-II' : datum.label == 54 ? 'Dengue 3-III' : datum.label == 57 ? 'Dengue 4-I' : datum.label == 58 ? 'Dengue 4-II' : datum.label == 64 ? 'Enterovirus D68' : 'Norovirus GII.4'\", \"tickCount\": 26, \"values\": [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 43, 44, 45, 48, 49, 50, 53, 54, 57, 58, 64, 67]}, \"field\": \"xtick_pos\", \"title\": \"\", \"type\": \"quantitative\"}, \"y\": {\"axis\": {\"format\": \".1f\", \"tickMinStep\": 1}, \"field\": \"adaptive_subs_per_codon_per_year\", \"title\": \"\", \"type\": \"quantitative\"}}, \"height\": 57.96833374243101, \"title\": \"Polymerase\", \"transform\": [{\"calculate\": \"('https://blab.github.io/atlas-of-viral-adaptation/' + datum.url_ending)\", \"as\": \"url\"}], \"width\": 800}, {\"data\": {\"name\": \"data-06a554f865048b667d6c6c7db9d5e2db\"}, \"mark\": \"errorbar\", \"encoding\": {\"color\": {\"field\": \"legible_name\", \"legend\": null, \"scale\": {\"domain\": [\"Influenza B/Vic\", \"229E\", \"Dengue 2-C\", \"Dengue 1-IV\", \"Influenza A/H3N2\", \"Lassa-A\", \"RSV-B\", \"RSV-A\", \"Dengue 3-II\", \"Dengue 4-I\", \"OC43-A\", \"Dengue 4-II\", \"Enterovirus D68\", \"Dengue 1-V\", \"Dengue 3-III\", \"Dengue 2-AI\", \"Influenza A/H1N1pdm\", \"Dengue 2-AA\", \"Mumps\", \"Lassa-B\", \"Parainfluenza-1\", \"Norovirus GII.4\", \"Parainfluenza-3\", \"NL63\", \"Influenza B/Yam\", \"Measles\", \"Dengue 1-III\", \"Dengue 1-I\"], \"range\": [\"#E06E85\", \"#76C7BE\", \"#2cccf2\", \"#0db0d6\", \"#9F2A44\", \"#9cc479\", \"#6C7593\", \"#0B194C\", \"#76def7\", \"#adebfa\", \"#208288\", \"#adebfa\", \"#a373a5\", \"#0db0d6\", \"#76def7\", \"#2cccf2\", \"#CC415A\", \"#2cccf2\", \"#eb520d\", \"#c4dcaf\", \"#f58451\", \"#d0b9d2\", \"#f79e76\", \"#BBE4D1\", \"#ED9AB0\", \"#F5AD52\", \"#0db0d6\", \"#0db0d6\"]}, \"type\": \"nominal\"}, \"tooltip\": {\"value\": null}, \"x\": {\"axis\": {\"labels\": false}, \"field\": \"xtick_pos\", \"sort\": [\"Measles\", \"Mumps\", \"Parainfluenza-1\", \"Parainfluenza-3\", \"Influenza A/H3N2\", \"Influenza A/H1N1pdm\", \"Influenza B/Vic\", \"Influenza B/Yam\", \"OC43-A\", \"229E\", \"NL63\", \"RSV-A\", \"RSV-B\", \"Dengue 1-I\", \"Dengue 1-III\", \"Dengue 1-IV\", \"Dengue 1-V\", \"Dengue 2-AA\", \"Dengue 2-AI\", \"Dengue 2-C\", \"Dengue 3-II\", \"Dengue 3-III\", \"Dengue 4-I\", \"Dengue 4-II\", \"Enterovirus D68\", \"Norovirus GII.4\"], \"title\": \"\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"lower_95ci\", \"title\": \"\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"upper_95ci\"}}, \"height\": 57.96833374243101, \"width\": 800}, {\"data\": {\"name\": \"data-0209ec4f059fa6853fca90f6e38626c5\"}, \"mark\": {\"type\": \"rule\", \"strokeDash\": [3, 5]}, \"encoding\": {\"x\": {\"field\": \"x\", \"type\": \"quantitative\"}}}], \"data\": {\"name\": \"data-06a554f865048b667d6c6c7db9d5e2db\"}}, {\"layer\": [{\"data\": {\"name\": \"data-e5bd9ca9abcbf571067ef56c644e2ec6\"}, \"mark\": {\"type\": \"point\", \"filled\": true, \"opacity\": 1, \"size\": 200}, \"encoding\": {\"color\": {\"field\": \"legible_name\", \"legend\": null, \"scale\": {\"domain\": [\"Influenza B/Vic\", \"229E\", \"Dengue 2-C\", \"Dengue 1-IV\", \"Influenza A/H3N2\", \"Lassa-A\", \"RSV-B\", \"RSV-A\", \"Dengue 3-II\", \"Dengue 4-I\", \"OC43-A\", \"Dengue 4-II\", \"Enterovirus D68\", \"Dengue 1-V\", \"Dengue 3-III\", \"Dengue 2-AI\", \"Influenza A/H1N1pdm\", \"Dengue 2-AA\", \"Mumps\", \"Lassa-B\", \"Parainfluenza-1\", \"Norovirus GII.4\", \"Parainfluenza-3\", \"NL63\", \"Influenza B/Yam\", \"Measles\", \"Dengue 1-III\", \"Dengue 1-I\"], \"range\": [\"#E06E85\", \"#76C7BE\", \"#2cccf2\", \"#0db0d6\", \"#9F2A44\", \"#9cc479\", \"#6C7593\", \"#0B194C\", \"#76def7\", \"#adebfa\", \"#208288\", \"#adebfa\", \"#a373a5\", \"#0db0d6\", \"#76def7\", \"#2cccf2\", \"#CC415A\", \"#2cccf2\", \"#eb520d\", \"#c4dcaf\", \"#f58451\", \"#d0b9d2\", \"#f79e76\", \"#BBE4D1\", \"#ED9AB0\", \"#F5AD52\", \"#0db0d6\", \"#0db0d6\"]}, \"type\": \"nominal\"}, \"href\": {\"field\": \"url\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"legible_name\", \"title\": \"Virus\", \"type\": \"nominal\"}, {\"field\": \"virus_family\", \"title\": \"Family\", \"type\": \"nominal\"}, {\"field\": \"genome_type\", \"title\": \"Genome Type\", \"type\": \"nominal\"}, {\"field\": \"enveloped\", \"title\": \"Enveloped\", \"type\": \"nominal\"}, {\"field\": \"legible_gene_name\", \"title\": \"Protein/Subunit\", \"type\": \"nominal\"}, {\"field\": \"legible_rate_percodon\", \"title\": \"Rate\", \"type\": \"nominal\"}], \"x\": {\"axis\": {\"labelExpr\": \"datum.label == 3 ? 'Measles' : datum.label == 6 ? 'Mumps' : datum.label == 9 ? 'Parainfluenza-1' : datum.label == 12 ? 'Parainfluenza-3' : datum.label == 15 ? 'Influenza A/H3N2' : datum.label == 18 ? 'Influenza A/H1N1pdm' : datum.label == 21 ? 'Influenza B/Vic' : datum.label == 24 ? 'Influenza B/Yam' : datum.label == 27 ? 'OC43-A' : datum.label == 30 ? '229E' : datum.label == 33 ? 'NL63' : datum.label == 36 ? 'RSV-A' : datum.label == 39 ? 'RSV-B' : datum.label == 42 ? 'Dengue 1-I' : datum.label == 43 ? 'Dengue 1-III' : datum.label == 44 ? 'Dengue 1-IV' : datum.label == 45 ? 'Dengue 1-V' : datum.label == 48 ? 'Dengue 2-AA' : datum.label == 49 ? 'Dengue 2-AI' : datum.label == 50 ? 'Dengue 2-C' : datum.label == 53 ? 'Dengue 3-II' : datum.label == 54 ? 'Dengue 3-III' : datum.label == 57 ? 'Dengue 4-I' : datum.label == 58 ? 'Dengue 4-II' : datum.label == 64 ? 'Enterovirus D68' : 'Norovirus GII.4'\", \"tickCount\": 26, \"values\": [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 43, 44, 45, 48, 49, 50, 53, 54, 57, 58, 64, 67]}, \"field\": \"xtick_pos\", \"title\": \"\", \"type\": \"quantitative\"}, \"y\": {\"axis\": {\"format\": \".1f\"}, \"field\": \"adaptive_subs_per_codon_per_year\", \"title\": \"\", \"type\": \"quantitative\"}}, \"height\": 250, \"title\": \"Receptor-Binding\", \"transform\": [{\"calculate\": \"('https://blab.github.io/atlas-of-viral-adaptation/' + datum.url_ending)\", \"as\": \"url\"}], \"width\": 800}, {\"data\": {\"name\": \"data-e5bd9ca9abcbf571067ef56c644e2ec6\"}, \"mark\": \"errorbar\", \"encoding\": {\"color\": {\"field\": \"legible_name\", \"legend\": null, \"scale\": {\"domain\": [\"Influenza B/Vic\", \"229E\", \"Dengue 2-C\", \"Dengue 1-IV\", \"Influenza A/H3N2\", \"Lassa-A\", \"RSV-B\", \"RSV-A\", \"Dengue 3-II\", \"Dengue 4-I\", \"OC43-A\", \"Dengue 4-II\", \"Enterovirus D68\", \"Dengue 1-V\", \"Dengue 3-III\", \"Dengue 2-AI\", \"Influenza A/H1N1pdm\", \"Dengue 2-AA\", \"Mumps\", \"Lassa-B\", \"Parainfluenza-1\", \"Norovirus GII.4\", \"Parainfluenza-3\", \"NL63\", \"Influenza B/Yam\", \"Measles\", \"Dengue 1-III\", \"Dengue 1-I\"], \"range\": [\"#E06E85\", \"#76C7BE\", \"#2cccf2\", \"#0db0d6\", \"#9F2A44\", \"#9cc479\", \"#6C7593\", \"#0B194C\", \"#76def7\", \"#adebfa\", \"#208288\", \"#adebfa\", \"#a373a5\", \"#0db0d6\", \"#76def7\", \"#2cccf2\", \"#CC415A\", \"#2cccf2\", \"#eb520d\", \"#c4dcaf\", \"#f58451\", \"#d0b9d2\", \"#f79e76\", \"#BBE4D1\", \"#ED9AB0\", \"#F5AD52\", \"#0db0d6\", \"#0db0d6\"]}, \"type\": \"nominal\"}, \"tooltip\": {\"value\": null}, \"x\": {\"field\": \"xtick_pos\", \"sort\": [\"Measles\", \"Mumps\", \"Parainfluenza-1\", \"Parainfluenza-3\", \"Influenza A/H3N2\", \"Influenza A/H1N1pdm\", \"Influenza B/Vic\", \"Influenza B/Yam\", \"OC43-A\", \"229E\", \"NL63\", \"RSV-A\", \"RSV-B\", \"Dengue 1-I\", \"Dengue 1-III\", \"Dengue 1-IV\", \"Dengue 1-V\", \"Dengue 2-AA\", \"Dengue 2-AI\", \"Dengue 2-C\", \"Dengue 3-II\", \"Dengue 3-III\", \"Dengue 4-I\", \"Dengue 4-II\", \"Enterovirus D68\", \"Norovirus GII.4\"], \"title\": \"\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"lower_95ci\", \"title\": \"\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"upper_95ci\"}}, \"height\": 250, \"width\": 800}, {\"data\": {\"name\": \"data-0209ec4f059fa6853fca90f6e38626c5\"}, \"mark\": {\"type\": \"rule\", \"strokeDash\": [3, 5]}, \"encoding\": {\"x\": {\"field\": \"x\", \"type\": \"quantitative\"}}}, {\"data\": {\"name\": \"data-0209ec4f059fa6853fca90f6e38626c5\"}, \"mark\": {\"type\": \"text\", \"align\": \"right\", \"color\": \"#6e6e6e\", \"dx\": -10, \"dy\": -120}, \"encoding\": {\"text\": {\"value\": [\"enveloped\", \"viruses\"]}, \"x\": {\"field\": \"x\", \"type\": \"quantitative\"}}}, {\"data\": {\"name\": \"data-0209ec4f059fa6853fca90f6e38626c5\"}, \"mark\": {\"type\": \"text\", \"align\": \"left\", \"color\": \"#6e6e6e\", \"dx\": 10, \"dy\": -120}, \"encoding\": {\"text\": {\"value\": [\"non-enveloped\", \"viruses\"]}, \"x\": {\"field\": \"x\", \"type\": \"quantitative\"}}}], \"data\": {\"name\": \"data-e5bd9ca9abcbf571067ef56c644e2ec6\"}}], \"spacing\": 10}], \"resolve\": {\"scale\": {\"x\": \"independent\", \"y\": \"independent\"}}, \"spacing\": 0, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"empty\": [{}], \"data-06a554f865048b667d6c6c7db9d5e2db\": [{\"virus\": \"measles\", \"subtype\": null, \"virus_and_subtype\": \"measles\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0696477489789021, \"lower_95ci\": 0.0019629917949619, \"upper_95ci\": 0.1477218174503109, \"ci\": \"[0.00196299179496196, 0.14772181745031096]\", \"len_in_codons\": 2184, \"adaptive_muts_per_year\": 0.1521106837699223, \"lower_95ci_mutspergene\": 0.0042871740801969, \"upper_95ci_mutspergene\": 0.3226244493114791, \"legible_name\": \"Measles\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#F5AD52\", \"index\": 0, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"measles\", \"xtick_pos\": 3, \"legible_rate_percodon\": \"0.07\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.15 muts per year\"}, {\"virus\": \"mumps\", \"subtype\": null, \"virus_and_subtype\": \"mumps\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": -0.0036394625647537, \"lower_95ci\": -0.0116868160093167, \"upper_95ci\": 0.016236636926292, \"ci\": \"[-0.011686816009316748, 0.016236636926292045]\", \"len_in_codons\": 2262, \"adaptive_muts_per_year\": -0.008232464321473, \"lower_95ci_mutspergene\": -0.0264355778130744, \"upper_95ci_mutspergene\": 0.0367272727272726, \"legible_name\": \"Mumps\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#eb520d\", \"index\": 3, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"mumps\", \"xtick_pos\": 6, \"legible_rate_percodon\": \"-0.0\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"-0.01 muts per year\"}, {\"virus\": \"parainfluenza\", \"subtype\": \"1\", \"virus_and_subtype\": \"parainfluenza_1\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1245544489235666, \"lower_95ci\": 0.0213585194689124, \"upper_95ci\": 0.2496937595953217, \"ci\": \"[0.021358519468912436, 0.24969375959532172]\", \"len_in_codons\": 2224, \"adaptive_muts_per_year\": 0.2770090944060122, \"lower_95ci_mutspergene\": 0.0475013472988612, \"upper_95ci_mutspergene\": 0.5553189213399955, \"legible_name\": \"Parainfluenza-1\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#f58451\", \"index\": 21, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"parainfluenza_1\", \"xtick_pos\": 9, \"legible_rate_percodon\": \"0.12\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.28 muts per year\"}, {\"virus\": \"parainfluenza\", \"subtype\": \"3\", \"virus_and_subtype\": \"parainfluenza_3\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0168440508390062, \"lower_95ci\": 0.0, \"upper_95ci\": 0.0518289545156088, \"ci\": \"[0.0, 0.05182895451560882]\", \"len_in_codons\": 2234, \"adaptive_muts_per_year\": 0.0376296095743399, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1157858843878701, \"legible_name\": \"Parainfluenza-3\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#f79e76\", \"index\": 21, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"parainfluenza_3\", \"xtick_pos\": 12, \"legible_rate_percodon\": \"0.02\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.04 muts per year\"}, {\"virus\": \"h3n2\", \"subtype\": null, \"virus_and_subtype\": \"h3n2\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.3804195219421517, \"lower_95ci\": 0.2019456795663166, \"upper_95ci\": 0.587181685626881, \"ci\": \"[0.2019456795663166, 0.587181685626881]\", \"len_in_codons\": 758, \"adaptive_muts_per_year\": 0.288357997632151, \"lower_95ci_mutspergene\": 0.153074825111268, \"upper_95ci_mutspergene\": 0.4450837177051758, \"legible_name\": \"Influenza A/H3N2\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#9F2A44\", \"index\": 24, \"legible_gene_name\": \"PB1 (Polymerase)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"h3n2\", \"xtick_pos\": 15, \"legible_rate_percodon\": \"0.38\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.29 muts per year\"}, {\"virus\": \"h1n1pdm\", \"subtype\": null, \"virus_and_subtype\": \"h1n1pdm\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2895886158625782, \"lower_95ci\": -0.001560244403758, \"upper_95ci\": 1.132737866649248, \"ci\": \"[-0.0015602444037580434, 1.132737866649248]\", \"len_in_codons\": 758, \"adaptive_muts_per_year\": 0.2195081708238342, \"lower_95ci_mutspergene\": -0.0011826652580485, \"upper_95ci_mutspergene\": 0.85861530292013, \"legible_name\": \"Influenza A/H1N1pdm\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#CC415A\", \"index\": 6, \"legible_gene_name\": \"PB1 (Polymerase)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"h1n1pdm\", \"xtick_pos\": 18, \"legible_rate_percodon\": \"0.29\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.22 muts per year\"}, {\"virus\": \"vic\", \"subtype\": null, \"virus_and_subtype\": \"vic\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1515692519238777, \"lower_95ci\": 0.0023655998857976, \"upper_95ci\": 0.315418668570784, \"ci\": \"[0.0023655998857976504, 0.31541866857078404]\", \"len_in_codons\": 753, \"adaptive_muts_per_year\": 0.1141316466986799, \"lower_95ci_mutspergene\": 0.0017812967140056, \"upper_95ci_mutspergene\": 0.2375102574338004, \"legible_name\": \"Influenza B/Vic\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#E06E85\", \"index\": 15, \"legible_gene_name\": \"PB1 (Polymerase)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"vic\", \"xtick_pos\": 21, \"legible_rate_percodon\": \"0.15\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"yam\", \"subtype\": null, \"virus_and_subtype\": \"yam\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1049109846029593, \"lower_95ci\": -0.0070505300278014, \"upper_95ci\": 0.2730279027767181, \"ci\": \"[-0.00705053002780142, 0.2730279027767181]\", \"len_in_codons\": 753, \"adaptive_muts_per_year\": 0.0789979714060283, \"lower_95ci_mutspergene\": -0.0053090491109344, \"upper_95ci_mutspergene\": 0.2055900107908687, \"legible_name\": \"Influenza B/Yam\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#ED9AB0\", \"index\": 9, \"legible_gene_name\": \"PB1 (Polymerase)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"yam\", \"xtick_pos\": 24, \"legible_rate_percodon\": \"0.1\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.08 muts per year\"}, {\"virus\": \"oc43\", \"subtype\": \"a\", \"virus_and_subtype\": \"oc43_a\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0368813805325201, \"lower_95ci\": 0.0, \"upper_95ci\": 0.1475255221300805, \"ci\": \"[0.0, 0.14752552213008055]\", \"len_in_codons\": 928, \"adaptive_muts_per_year\": 0.0342259211341786, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1369036845367147, \"legible_name\": \"OC43-A\", \"virus_family\": \"Coronavirus\", \"color\": \"#208288\", \"index\": 36, \"legible_gene_name\": \"RDRP (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"oc43_a\", \"xtick_pos\": 27, \"legible_rate_percodon\": \"0.04\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.03 muts per year\"}, {\"virus\": \"229e\", \"subtype\": null, \"virus_and_subtype\": \"229e\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0369532263703259, \"lower_95ci\": 0.0, \"upper_95ci\": 0.1349980102763556, \"ci\": \"[0.0, 0.13499801027635566]\", \"len_in_codons\": 927, \"adaptive_muts_per_year\": 0.0342556408452921, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1251431555261817, \"legible_name\": \"229E\", \"virus_family\": \"Coronavirus\", \"color\": \"#76C7BE\", \"index\": 18, \"legible_gene_name\": \"RDRP (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"229e\", \"xtick_pos\": 30, \"legible_rate_percodon\": \"0.04\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.03 muts per year\"}, {\"virus\": \"nl63\", \"subtype\": null, \"virus_and_subtype\": \"nl63\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0, \"lower_95ci\": 0.0, \"upper_95ci\": 0.0, \"ci\": \"[0.0, 0.0]\", \"len_in_codons\": 927, \"adaptive_muts_per_year\": 0.0, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.0, \"legible_name\": \"NL63\", \"virus_family\": \"Coronavirus\", \"color\": \"#BBE4D1\", \"index\": 27, \"legible_gene_name\": \"RDRP (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"nl63\", \"xtick_pos\": 33, \"legible_rate_percodon\": \"0.0\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.0 muts per year\"}, {\"virus\": \"rsv\", \"subtype\": \"A\", \"virus_and_subtype\": \"rsv_A\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1791526081703916, \"lower_95ci\": 0.0734714909962797, \"upper_95ci\": 0.2786727160334553, \"ci\": \"[0.07347149099627973, 0.27867271603345534]\", \"len_in_codons\": 2166, \"adaptive_muts_per_year\": 0.3880445492970683, \"lower_95ci_mutspergene\": 0.1591392494979418, \"upper_95ci_mutspergene\": 0.6036051029284643, \"legible_name\": \"RSV-A\", \"virus_family\": \"Pneumovirus\", \"color\": \"#0B194C\", \"index\": 39, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"rsv_a\", \"xtick_pos\": 36, \"legible_rate_percodon\": \"0.18\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.39 muts per year\"}, {\"virus\": \"rsv\", \"subtype\": \"B\", \"virus_and_subtype\": \"rsv_B\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.227456322682962, \"lower_95ci\": 0.0999231081717831, \"upper_95ci\": 0.325143188036216, \"ci\": \"[0.09992310817178313, 0.32514318803621606]\", \"len_in_codons\": 2166, \"adaptive_muts_per_year\": 0.4926703949312957, \"lower_95ci_mutspergene\": 0.2164334523000822, \"upper_95ci_mutspergene\": 0.704260145286444, \"legible_name\": \"RSV-B\", \"virus_family\": \"Pneumovirus\", \"color\": \"#6C7593\", \"index\": 39, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"rsv_b\", \"xtick_pos\": 39, \"legible_rate_percodon\": \"0.23\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.49 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_I\", \"virus_and_subtype\": \"dengue_denv1_I\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2548158661827924, \"lower_95ci\": 0.0359724661139091, \"upper_95ci\": 0.5927314906874545, \"ci\": \"[0.03597246611390916, 0.5927314906874545]\", \"len_in_codons\": 899, \"adaptive_muts_per_year\": 0.2290794636983303, \"lower_95ci_mutspergene\": 0.0323392470364043, \"upper_95ci_mutspergene\": 0.5328656101280216, \"legible_name\": \"Dengue 1-I\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_i\", \"xtick_pos\": 42, \"legible_rate_percodon\": \"0.25\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.23 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_III\", \"virus_and_subtype\": \"dengue_denv1_III\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0557099382380511, \"lower_95ci\": 0.0, \"upper_95ci\": 0.1650902814782644, \"ci\": \"[0.0, 0.16509028147826446]\", \"len_in_codons\": 899, \"adaptive_muts_per_year\": 0.050083234476008, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1484161630489597, \"legible_name\": \"Dengue 1-III\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_iii\", \"xtick_pos\": 43, \"legible_rate_percodon\": \"0.06\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.05 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_IV\", \"virus_and_subtype\": \"dengue_denv1_IV\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1536898027582185, \"lower_95ci\": 0.0093388951033292, \"upper_95ci\": 0.3542488014812919, \"ci\": \"[0.009338895103329251, 0.3542488014812919]\", \"len_in_codons\": 899, \"adaptive_muts_per_year\": 0.1381671326796384, \"lower_95ci_mutspergene\": 0.0083956666978929, \"upper_95ci_mutspergene\": 0.3184696725316814, \"legible_name\": \"Dengue 1-IV\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_iv\", \"xtick_pos\": 44, \"legible_rate_percodon\": \"0.15\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.14 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_V\", \"virus_and_subtype\": \"dengue_denv1_V\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1207176423891461, \"lower_95ci\": 0.0, \"upper_95ci\": 0.2506198310284858, \"ci\": \"[0.0, 0.2506198310284858]\", \"len_in_codons\": 899, \"adaptive_muts_per_year\": 0.1085251605078424, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.2253072280946087, \"legible_name\": \"Dengue 1-V\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_v\", \"xtick_pos\": 45, \"legible_rate_percodon\": \"0.12\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_AA\", \"virus_and_subtype\": \"dengue_denv2_AA\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2300077187985516, \"lower_95ci\": 0.0820793317286246, \"upper_95ci\": 0.4116735028777324, \"ci\": \"[0.08207933172862468, 0.41167350287773247]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.2070069469186964, \"lower_95ci_mutspergene\": 0.0738713985557622, \"upper_95ci_mutspergene\": 0.3705061525899592, \"legible_name\": \"Dengue 2-AA\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_aa\", \"xtick_pos\": 48, \"legible_rate_percodon\": \"0.23\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.21 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_AI\", \"virus_and_subtype\": \"dengue_denv2_AI\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1557132167090561, \"lower_95ci\": 0.0107601070156147, \"upper_95ci\": 0.3535455238972207, \"ci\": \"[0.010760107015614785, 0.35354552389722077]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.1401418950381505, \"lower_95ci_mutspergene\": 0.0096840963140533, \"upper_95ci_mutspergene\": 0.3181909715074987, \"legible_name\": \"Dengue 2-AI\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_ai\", \"xtick_pos\": 49, \"legible_rate_percodon\": \"0.16\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.14 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_C\", \"virus_and_subtype\": \"dengue_denv2_C\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": -0.0617916284955231, \"lower_95ci\": -0.1282134695879142, \"upper_95ci\": -0.0126741769450911, \"ci\": \"[-0.1282134695879142, -0.012674176945091181]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": -0.0556124656459708, \"lower_95ci_mutspergene\": -0.1153921226291227, \"upper_95ci_mutspergene\": -0.011406759250582, \"legible_name\": \"Dengue 2-C\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_c\", \"xtick_pos\": 50, \"legible_rate_percodon\": \"-0.06\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"-0.06 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv3_II\", \"virus_and_subtype\": \"dengue_denv3_II\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1270526611523327, \"lower_95ci\": 0.045667845029431, \"upper_95ci\": 0.204669909644794, \"ci\": \"[0.04566784502943106, 0.20466990964479406]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.1143473950370994, \"lower_95ci_mutspergene\": 0.0411010605264879, \"upper_95ci_mutspergene\": 0.1842029186803146, \"legible_name\": \"Dengue 3-II\", \"virus_family\": \"Flavivirus\", \"color\": \"#76def7\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv3_ii\", \"xtick_pos\": 53, \"legible_rate_percodon\": \"0.13\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv3_III\", \"virus_and_subtype\": \"dengue_denv3_III\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1179126365521947, \"lower_95ci\": 0.0204133578039758, \"upper_95ci\": 0.248946148912938, \"ci\": \"[0.02041335780397588, 0.24894614891293806]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.1061213728969753, \"lower_95ci_mutspergene\": 0.0183720220235782, \"upper_95ci_mutspergene\": 0.2240515340216442, \"legible_name\": \"Dengue 3-III\", \"virus_family\": \"Flavivirus\", \"color\": \"#76def7\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv3_iii\", \"xtick_pos\": 54, \"legible_rate_percodon\": \"0.12\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv4_I\", \"virus_and_subtype\": \"dengue_denv4_I\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2104987551233288, \"lower_95ci\": 0.0889963151806821, \"upper_95ci\": 0.3305517123186876, \"ci\": \"[0.08899631518068213, 0.3305517123186876]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.189448879610996, \"lower_95ci_mutspergene\": 0.0800966836626139, \"upper_95ci_mutspergene\": 0.2974965410868188, \"legible_name\": \"Dengue 4-I\", \"virus_family\": \"Flavivirus\", \"color\": \"#adebfa\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv4_i\", \"xtick_pos\": 57, \"legible_rate_percodon\": \"0.21\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.19 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv4_II\", \"virus_and_subtype\": \"dengue_denv4_II\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0723489159160415, \"lower_95ci\": 0.0286215625878451, \"upper_95ci\": 0.1252053439161044, \"ci\": \"[0.028621562587845113, 0.1252053439161044]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.0651140243244374, \"lower_95ci_mutspergene\": 0.0257594063290606, \"upper_95ci_mutspergene\": 0.1126848095244939, \"legible_name\": \"Dengue 4-II\", \"virus_family\": \"Flavivirus\", \"color\": \"#adebfa\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv4_ii\", \"xtick_pos\": 58, \"legible_rate_percodon\": \"0.07\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.07 muts per year\"}, {\"virus\": \"enterovirusd68\", \"subtype\": null, \"virus_and_subtype\": \"enterovirusd68\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.3421578702666206, \"lower_95ci\": 0.0312003072602813, \"upper_95ci\": 0.735949789390216, \"ci\": \"[0.031200307260281307, 0.735949789390216]\", \"len_in_codons\": 457, \"adaptive_muts_per_year\": 0.1563661467118456, \"lower_95ci_mutspergene\": 0.0142585404179485, \"upper_95ci_mutspergene\": 0.3363290537513287, \"legible_name\": \"Enterovirus D68\", \"virus_family\": \"Picornavirus\", \"color\": \"#a373a5\", \"index\": 30, \"legible_gene_name\": \"3D (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"False\", \"url_ending\": \"enterovirusd68\", \"xtick_pos\": 64, \"legible_rate_percodon\": \"0.34\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.16 muts per year\"}, {\"virus\": \"norovirus\", \"subtype\": null, \"virus_and_subtype\": \"norovirus\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2909630611667262, \"lower_95ci\": 0.0404426097320981, \"upper_95ci\": 0.5874570070327857, \"ci\": \"[0.040442609732098164, 0.5874570070327857]\", \"len_in_codons\": 510, \"adaptive_muts_per_year\": 0.1483911611950304, \"lower_95ci_mutspergene\": 0.02062573096337, \"upper_95ci_mutspergene\": 0.2996030735867207, \"legible_name\": \"Norovirus GII.4\", \"virus_family\": \"Caliciviridae\", \"color\": \"#d0b9d2\", \"index\": 12, \"legible_gene_name\": \"RDRP (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"False\", \"url_ending\": \"norovirus_gii4\", \"xtick_pos\": 67, \"legible_rate_percodon\": \"0.29\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.15 muts per year\"}], \"data-0209ec4f059fa6853fca90f6e38626c5\": [{\"x\": 61}], \"data-e5bd9ca9abcbf571067ef56c644e2ec6\": [{\"virus\": \"measles\", \"subtype\": null, \"virus_and_subtype\": \"measles\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.1488760781105891, \"lower_95ci\": 0.0, \"upper_95ci\": 0.464536965766201, \"ci\": \"[0.0, 0.46453696576620107]\", \"len_in_codons\": 618, \"adaptive_muts_per_year\": 0.0920054162723441, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.2870838448435123, \"legible_name\": \"Measles\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#F5AD52\", \"index\": 2, \"legible_gene_name\": \"H (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"measles\", \"xtick_pos\": 3, \"legible_rate_percodon\": \"0.15\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.09 muts per year\"}, {\"virus\": \"mumps\", \"subtype\": null, \"virus_and_subtype\": \"mumps\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.0177121918434842, \"lower_95ci\": 0.0, \"upper_95ci\": 0.0837561442164208, \"ci\": \"[0.0, 0.08375614421642087]\", \"len_in_codons\": 583, \"adaptive_muts_per_year\": 0.0103262078447513, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.0488298320781733, \"legible_name\": \"Mumps\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#eb520d\", \"index\": 5, \"legible_gene_name\": \"HN (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"mumps\", \"xtick_pos\": 6, \"legible_rate_percodon\": \"0.02\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.01 muts per year\"}, {\"virus\": \"parainfluenza\", \"subtype\": \"1\", \"virus_and_subtype\": \"parainfluenza_1\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.5301824829604939, \"lower_95ci\": -0.0807613335513118, \"upper_95ci\": 1.0682467463132916, \"ci\": \"[-0.08076133355131188, 1.0682467463132916]\", \"len_in_codons\": 576, \"adaptive_muts_per_year\": 0.3053851101852444, \"lower_95ci_mutspergene\": -0.0465185281255556, \"upper_95ci_mutspergene\": 0.6153101258764561, \"legible_name\": \"Parainfluenza-1\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#f58451\", \"index\": 23, \"legible_gene_name\": \"HN (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"parainfluenza_1\", \"xtick_pos\": 9, \"legible_rate_percodon\": \"0.53\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.31 muts per year\"}, {\"virus\": \"parainfluenza\", \"subtype\": \"3\", \"virus_and_subtype\": \"parainfluenza_3\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.0, \"lower_95ci\": 0.0, \"upper_95ci\": 0.0, \"ci\": \"[0.0, 0.0]\", \"len_in_codons\": 573, \"adaptive_muts_per_year\": 0.0, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.0, \"legible_name\": \"Parainfluenza-3\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#f79e76\", \"index\": 23, \"legible_gene_name\": \"HN (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"parainfluenza_3\", \"xtick_pos\": 12, \"legible_rate_percodon\": \"0.0\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.0 muts per year\"}, {\"virus\": \"h3n2\", \"subtype\": null, \"virus_and_subtype\": \"h3n2\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 3.968017826307325, \"lower_95ci\": 2.958853820205344, \"upper_95ci\": 4.99468539503025, \"ci\": \"[2.9588538202053436, 4.99468539503025]\", \"len_in_codons\": 329, \"adaptive_muts_per_year\": 1.3054778648551095, \"lower_95ci_mutspergene\": 0.973462906847558, \"upper_95ci_mutspergene\": 1.6432514949649522, \"legible_name\": \"Influenza A/H3N2\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#9F2A44\", \"index\": 26, \"legible_gene_name\": \"HA1 (Receptor Binding)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"h3n2\", \"xtick_pos\": 15, \"legible_rate_percodon\": \"3.97\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"1.31 muts per year\"}, {\"virus\": \"h1n1pdm\", \"subtype\": null, \"virus_and_subtype\": \"h1n1pdm\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 3.081395153594903, \"lower_95ci\": 1.3370908325016322, \"upper_95ci\": 5.264504569755848, \"ci\": \"[1.3370908325016322, 5.264504569755848]\", \"len_in_codons\": 327, \"adaptive_muts_per_year\": 1.0076162152255332, \"lower_95ci_mutspergene\": 0.4372287022280337, \"upper_95ci_mutspergene\": 1.7214929943101624, \"legible_name\": \"Influenza A/H1N1pdm\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#CC415A\", \"index\": 8, \"legible_gene_name\": \"HA1 (Receptor Binding)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"h1n1pdm\", \"xtick_pos\": 18, \"legible_rate_percodon\": \"3.08\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"1.01 muts per year\"}, {\"virus\": \"vic\", \"subtype\": null, \"virus_and_subtype\": \"vic\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.1534834193373809, \"lower_95ci\": 0.5371777048905576, \"upper_95ci\": 1.7054541798505731, \"ci\": \"[0.5371777048905576, 1.7054541798505731]\", \"len_in_codons\": 346, \"adaptive_muts_per_year\": 0.3991052630907338, \"lower_95ci_mutspergene\": 0.1858634858921329, \"upper_95ci_mutspergene\": 0.5900871462282983, \"legible_name\": \"Influenza B/Vic\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#E06E85\", \"index\": 17, \"legible_gene_name\": \"HA1 (Receptor Binding)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"vic\", \"xtick_pos\": 21, \"legible_rate_percodon\": \"1.15\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.4 muts per year\"}, {\"virus\": \"yam\", \"subtype\": null, \"virus_and_subtype\": \"yam\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.5005224136436983, \"lower_95ci\": 0.5579713024604996, \"upper_95ci\": 2.221057553942913, \"ci\": \"[0.5579713024604996, 2.221057553942913]\", \"len_in_codons\": 346, \"adaptive_muts_per_year\": 0.5191807551207196, \"lower_95ci_mutspergene\": 0.1930580706513328, \"upper_95ci_mutspergene\": 0.7684859136642479, \"legible_name\": \"Influenza B/Yam\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#ED9AB0\", \"index\": 11, \"legible_gene_name\": \"HA1 (Receptor Binding)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"yam\", \"xtick_pos\": 24, \"legible_rate_percodon\": \"1.5\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.52 muts per year\"}, {\"virus\": \"oc43\", \"subtype\": \"a\", \"virus_and_subtype\": \"oc43_a\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.866573224980902, \"lower_95ci\": 1.0689984441339817, \"upper_95ci\": 2.828605874053624, \"ci\": \"[1.0689984441339815, 2.828605874053624]\", \"len_in_codons\": 753, \"adaptive_muts_per_year\": 1.405529638410619, \"lower_95ci_mutspergene\": 0.804955828432888, \"upper_95ci_mutspergene\": 2.129940223162379, \"legible_name\": \"OC43-A\", \"virus_family\": \"Coronavirus\", \"color\": \"#208288\", \"index\": 38, \"legible_gene_name\": \"S1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"oc43_a\", \"xtick_pos\": 27, \"legible_rate_percodon\": \"1.87\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"1.41 muts per year\"}, {\"virus\": \"229e\", \"subtype\": null, \"virus_and_subtype\": \"229e\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.4851663127760144, \"lower_95ci\": 0.9834867527798624, \"upper_95ci\": 2.068365846820297, \"ci\": \"[0.9834867527798623, 2.0683658468202974]\", \"len_in_codons\": 545, \"adaptive_muts_per_year\": 0.8094156404629278, \"lower_95ci_mutspergene\": 0.5360002802650249, \"upper_95ci_mutspergene\": 1.127259386517062, \"legible_name\": \"229E\", \"virus_family\": \"Coronavirus\", \"color\": \"#76C7BE\", \"index\": 20, \"legible_gene_name\": \"S1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"229e\", \"xtick_pos\": 30, \"legible_rate_percodon\": \"1.49\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.81 muts per year\"}, {\"virus\": \"nl63\", \"subtype\": null, \"virus_and_subtype\": \"nl63\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.3233562876592263, \"lower_95ci\": 0.1399656586773711, \"upper_95ci\": 0.5437555140139991, \"ci\": \"[0.13996565867737112, 0.5437555140139991]\", \"len_in_codons\": 734, \"adaptive_muts_per_year\": 0.2373435151418721, \"lower_95ci_mutspergene\": 0.1027347934691904, \"upper_95ci_mutspergene\": 0.3991165472862754, \"legible_name\": \"NL63\", \"virus_family\": \"Coronavirus\", \"color\": \"#BBE4D1\", \"index\": 29, \"legible_gene_name\": \"S1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"nl63\", \"xtick_pos\": 33, \"legible_rate_percodon\": \"0.32\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.24 muts per year\"}, {\"virus\": \"rsv\", \"subtype\": \"A\", \"virus_and_subtype\": \"rsv_A\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.9300665866808224, \"lower_95ci\": 1.1735832138549698, \"upper_95ci\": 2.878035332770765, \"ci\": \"[1.1735832138549698, 2.878035332770765]\", \"len_in_codons\": 298, \"adaptive_muts_per_year\": 0.5751598428308852, \"lower_95ci_mutspergene\": 0.349727797728781, \"upper_95ci_mutspergene\": 0.8576545291656879, \"legible_name\": \"RSV-A\", \"virus_family\": \"Pneumovirus\", \"color\": \"#0B194C\", \"index\": 41, \"legible_gene_name\": \"G (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"rsv_a\", \"xtick_pos\": 36, \"legible_rate_percodon\": \"1.93\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.58 muts per year\"}, {\"virus\": \"rsv\", \"subtype\": \"B\", \"virus_and_subtype\": \"rsv_B\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 2.1439093828606683, \"lower_95ci\": 1.2801227304766734, \"upper_95ci\": 3.3767220257559565, \"ci\": \"[1.2801227304766734, 3.3767220257559565]\", \"len_in_codons\": 300, \"adaptive_muts_per_year\": 0.6431728148582004, \"lower_95ci_mutspergene\": 0.384036819143002, \"upper_95ci_mutspergene\": 1.0130166077267868, \"legible_name\": \"RSV-B\", \"virus_family\": \"Pneumovirus\", \"color\": \"#6C7593\", \"index\": 41, \"legible_gene_name\": \"G (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"rsv_b\", \"xtick_pos\": 39, \"legible_rate_percodon\": \"2.14\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.64 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_I\", \"virus_and_subtype\": \"dengue_denv1_I\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.2772247014216659, \"lower_95ci\": 0.0480079828890349, \"upper_95ci\": 0.6077131469207415, \"ci\": \"[0.04800798288903498, 0.6077131469207415]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.1372262272037246, \"lower_95ci_mutspergene\": 0.0237639515300723, \"upper_95ci_mutspergene\": 0.300818007725767, \"legible_name\": \"Dengue 1-I\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_i\", \"xtick_pos\": 42, \"legible_rate_percodon\": \"0.28\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.14 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_III\", \"virus_and_subtype\": \"dengue_denv1_III\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.0840528138476497, \"lower_95ci\": 0.0, \"upper_95ci\": 0.2912781302534159, \"ci\": \"[0.0, 0.2912781302534159]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.0416061428545866, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1441826744754408, \"legible_name\": \"Dengue 1-III\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_iii\", \"xtick_pos\": 43, \"legible_rate_percodon\": \"0.08\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.04 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_IV\", \"virus_and_subtype\": \"dengue_denv1_IV\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.3275670161682112, \"lower_95ci\": 0.1260494030967501, \"upper_95ci\": 0.6447779427586959, \"ci\": \"[0.12604940309675014, 0.6447779427586959]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.1621456730032646, \"lower_95ci_mutspergene\": 0.0623944545328913, \"upper_95ci_mutspergene\": 0.3191650816655545, \"legible_name\": \"Dengue 1-IV\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_iv\", \"xtick_pos\": 44, \"legible_rate_percodon\": \"0.33\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.16 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_V\", \"virus_and_subtype\": \"dengue_denv1_V\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": -0.0095111020473242, \"lower_95ci\": -0.0304944848967084, \"upper_95ci\": 0.0, \"ci\": \"[-0.03049448489670842, 0.0]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": -0.0047079955134255, \"lower_95ci_mutspergene\": -0.0150947700238706, \"upper_95ci_mutspergene\": 0.0, \"legible_name\": \"Dengue 1-V\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_v\", \"xtick_pos\": 45, \"legible_rate_percodon\": \"-0.01\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"-0.0 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_AA\", \"virus_and_subtype\": \"dengue_denv2_AA\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.3356339849242723, \"lower_95ci\": 0.0342444878870831, \"upper_95ci\": 0.8756110481375577, \"ci\": \"[0.03424448788708317, 0.8756110481375577]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.1661388225375148, \"lower_95ci_mutspergene\": 0.0169510215041061, \"upper_95ci_mutspergene\": 0.4334274688280911, \"legible_name\": \"Dengue 2-AA\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_aa\", \"xtick_pos\": 48, \"legible_rate_percodon\": \"0.34\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.17 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_AI\", \"virus_and_subtype\": \"dengue_denv2_AI\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.231380061824413, \"lower_95ci\": 0.0231912292530623, \"upper_95ci\": 0.466973072382025, \"ci\": \"[0.023191229253062343, 0.466973072382025]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.1145331306030844, \"lower_95ci_mutspergene\": 0.0114796584802658, \"upper_95ci_mutspergene\": 0.2311516708291023, \"legible_name\": \"Dengue 2-AI\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_ai\", \"xtick_pos\": 49, \"legible_rate_percodon\": \"0.23\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_C\", \"virus_and_subtype\": \"dengue_denv2_C\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": -0.0643750628828253, \"lower_95ci\": -0.1735995417057223, \"upper_95ci\": 0.081578958446469, \"ci\": \"[-0.17359954170572234, 0.08157895844646901]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": -0.0318656561269985, \"lower_95ci_mutspergene\": -0.0859317731443325, \"upper_95ci_mutspergene\": 0.0403815844310021, \"legible_name\": \"Dengue 2-C\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_c\", \"xtick_pos\": 50, \"legible_rate_percodon\": \"-0.06\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"-0.03 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv3_II\", \"virus_and_subtype\": \"dengue_denv3_II\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.1910357962579551, \"lower_95ci\": 0.0827209609842614, \"upper_95ci\": 0.3886853493722831, \"ci\": \"[0.08272096098426143, 0.38868534937228316]\", \"len_in_codons\": 493, \"adaptive_muts_per_year\": 0.0941806475551718, \"lower_95ci_mutspergene\": 0.0407814337652408, \"upper_95ci_mutspergene\": 0.1916218772405356, \"legible_name\": \"Dengue 3-II\", \"virus_family\": \"Flavivirus\", \"color\": \"#76def7\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv3_ii\", \"xtick_pos\": 53, \"legible_rate_percodon\": \"0.19\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.09 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv3_III\", \"virus_and_subtype\": \"dengue_denv3_III\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.0317617470437703, \"lower_95ci\": -0.0001068335658367, \"upper_95ci\": 0.1281931768446245, \"ci\": \"[-0.0001068335658367902, 0.1281931768446245]\", \"len_in_codons\": 493, \"adaptive_muts_per_year\": 0.0156585412925788, \"lower_95ci_mutspergene\": -5.2668947957537565e-05, \"upper_95ci_mutspergene\": 0.0631992361843998, \"legible_name\": \"Dengue 3-III\", \"virus_family\": \"Flavivirus\", \"color\": \"#76def7\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv3_iii\", \"xtick_pos\": 54, \"legible_rate_percodon\": \"0.03\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.02 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv4_I\", \"virus_and_subtype\": \"dengue_denv4_I\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.4061259877770878, \"lower_95ci\": 0.1871819341848133, \"upper_95ci\": 0.6295197691163875, \"ci\": \"[0.1871819341848133, 0.6295197691163875]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.2010323639496584, \"lower_95ci_mutspergene\": 0.0926550574214825, \"upper_95ci_mutspergene\": 0.3116122857126118, \"legible_name\": \"Dengue 4-I\", \"virus_family\": \"Flavivirus\", \"color\": \"#adebfa\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv4_i\", \"xtick_pos\": 57, \"legible_rate_percodon\": \"0.41\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.2 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv4_II\", \"virus_and_subtype\": \"dengue_denv4_II\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.1246085081342652, \"lower_95ci\": 0.0161460481464703, \"upper_95ci\": 0.3103463519891937, \"ci\": \"[0.01614604814647036, 0.31034635198919375]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.0616812115264612, \"lower_95ci_mutspergene\": 0.0079922938325028, \"upper_95ci_mutspergene\": 0.1536214442346509, \"legible_name\": \"Dengue 4-II\", \"virus_family\": \"Flavivirus\", \"color\": \"#adebfa\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv4_ii\", \"xtick_pos\": 58, \"legible_rate_percodon\": \"0.12\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.06 muts per year\"}, {\"virus\": \"enterovirusd68\", \"subtype\": null, \"virus_and_subtype\": \"enterovirusd68\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.517049084058603, \"lower_95ci\": 0.7582093757458963, \"upper_95ci\": 2.327216254157004, \"ci\": \"[0.7582093757458963, 2.327216254157004]\", \"len_in_codons\": 309, \"adaptive_muts_per_year\": 0.4687681669741083, \"lower_95ci_mutspergene\": 0.2342866971054819, \"upper_95ci_mutspergene\": 0.7191098225345142, \"legible_name\": \"Enterovirus D68\", \"virus_family\": \"Picornavirus\", \"color\": \"#a373a5\", \"index\": 32, \"legible_gene_name\": \"VP1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"False\", \"url_ending\": \"enterovirusd68\", \"xtick_pos\": 64, \"legible_rate_percodon\": \"1.52\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.47 muts per year\"}, {\"virus\": \"norovirus\", \"subtype\": null, \"virus_and_subtype\": \"norovirus\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.81000833263981, \"lower_95ci\": 1.1967182413852349, \"upper_95ci\": 2.309784031755586, \"ci\": \"[1.1967182413852349, 2.309784031755586]\", \"len_in_codons\": 541, \"adaptive_muts_per_year\": 0.9792145079581372, \"lower_95ci_mutspergene\": 0.647424568589412, \"upper_95ci_mutspergene\": 1.249593161179772, \"legible_name\": \"Norovirus GII.4\", \"virus_family\": \"Caliciviridae\", \"color\": \"#d0b9d2\", \"index\": 14, \"legible_gene_name\": \"VP1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"False\", \"url_ending\": \"norovirus_gii4\", \"xtick_pos\": 67, \"legible_rate_percodon\": \"1.81\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.98 muts per year\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.HConcatChart(...)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#set height and width of plots\n",
    "standard_width = 800\n",
    "standard_height = 250\n",
    "\n",
    "\n",
    "# generate the points\n",
    "points_p = alt.Chart(df_polymerase).transform_calculate(\n",
    "    url='https://blab.github.io/atlas-of-viral-adaptation/' + alt.datum.url_ending\n",
    ").mark_point(\n",
    "    filled=True,\n",
    "    size=200, \n",
    "    opacity=1\n",
    ").encode(\n",
    "    x=alt.X('xtick_pos:Q', title='', \n",
    "            axis=alt.Axis(values = all_xticks, labelExpr=label_expr_str, tickCount=len(all_xticks))),\n",
    "    y=alt.Y('adaptive_subs_per_codon_per_year:Q', title='', axis=alt.Axis(format=\".1f\", tickMinStep=1)),\n",
    "    color=alt.Color('legible_name', scale=alt.Scale(domain=dom, range=rng), legend=None),\n",
    "    href='url:N',\n",
    "    tooltip=[alt.Tooltip('legible_name:N', title='Virus'),\n",
    "             alt.Tooltip('virus_family:N', title='Family'),\n",
    "             alt.Tooltip('genome_type:N', title='Genome Type'),\n",
    "             alt.Tooltip('enveloped:N', title='Enveloped'),\n",
    "             alt.Tooltip('legible_gene_name:N', title='Protein/Subunit'), \n",
    "             alt.Tooltip('legible_rate_percodon', title='Rate')],\n",
    ").properties(\n",
    "    width=standard_width,\n",
    "    height=standard_height*p_relative_range, \n",
    "    title='Polymerase'\n",
    ")\n",
    "\n",
    "points_r = alt.Chart(df_receptorbinding).transform_calculate(\n",
    "    url='https://blab.github.io/atlas-of-viral-adaptation/' + alt.datum.url_ending\n",
    ").mark_point(\n",
    "    filled=True,\n",
    "    size=200, opacity=1,\n",
    ").encode(\n",
    "    x=alt.X('xtick_pos:Q', title='', \n",
    "            axis=alt.Axis(values = all_xticks, labelExpr=label_expr_str, tickCount=len(all_xticks))),\n",
    "    y=alt.Y('adaptive_subs_per_codon_per_year:Q', title='', axis=alt.Axis(format=\".1f\")),\n",
    "    color=alt.Color('legible_name', scale=alt.Scale(domain=dom, range=rng), legend=None),\n",
    "    href='url:N',\n",
    "    tooltip=[alt.Tooltip('legible_name:N', title='Virus'),\n",
    "             alt.Tooltip('virus_family:N', title='Family'),\n",
    "             alt.Tooltip('genome_type:N', title='Genome Type'),\n",
    "             alt.Tooltip('enveloped:N', title='Enveloped'),\n",
    "             alt.Tooltip('legible_gene_name:N', title='Protein/Subunit'), \n",
    "             alt.Tooltip('legible_rate_percodon', title='Rate')],\n",
    ").properties(\n",
    "    width=standard_width,\n",
    "    height=standard_height, \n",
    "    title='Receptor-Binding'\n",
    ")\n",
    "\n",
    "#generate the error bars\n",
    "errorbars_p = alt.Chart(df_polymerase).mark_errorbar().encode(\n",
    "    x=alt.X('xtick_pos:Q', sort=virus_order, title='', axis=alt.Axis(labels=False)),\n",
    "    y=alt.Y(\"lower_95ci\", title=''),\n",
    "    y2=\"upper_95ci\", \n",
    "    color=alt.Color('legible_name', scale=alt.Scale(domain=dom, range=rng), legend=None),\n",
    "    tooltip=alt.value(None)\n",
    ").properties(\n",
    "    width=standard_width,\n",
    "    height=standard_height*p_relative_range\n",
    ")\n",
    "\n",
    "\n",
    "errorbars_r = alt.Chart(df_receptorbinding).mark_errorbar().encode(\n",
    "    x=alt.X('xtick_pos:Q', sort=virus_order, title=''),\n",
    "    y=alt.Y(\"lower_95ci\", title=''),\n",
    "    y2=\"upper_95ci\",\n",
    "    color=alt.Color('legible_name', scale=alt.Scale(domain=dom, range=rng), legend=None),\n",
    "    tooltip=alt.value(None)\n",
    ").properties(\n",
    "    width=standard_width,\n",
    "    height=standard_height\n",
    ")\n",
    "\n",
    "#vertical line to divide enveloped from non-enveloped viruses\n",
    "vert_divider = alt.Chart(pd.DataFrame({'x': [enveloped_vert_divider]})).mark_rule(strokeDash=[3,5]).encode(x='x')\n",
    "enveloped_text = vert_divider.mark_text(\n",
    "        align='right',\n",
    "        color='#6e6e6e',\n",
    "        dy=-120,\n",
    "        dx=-10  # Nudges text to left so it doesn't appear on top of the bar. in pixels\n",
    ").encode(\n",
    "    text=alt.value(['enveloped', 'viruses']))\n",
    "nonenveloped_text = vert_divider.mark_text(\n",
    "        align='left',\n",
    "        color='#6e6e6e',\n",
    "        dy=-120,\n",
    "        dx=10  # Nudges text to left so it doesn't appear on top of the bar. in pixels\n",
    ").encode(\n",
    "    text=alt.value(['non-enveloped', 'viruses']))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#y-axis label for both plots\n",
    "text = alt.Chart().mark_text(\n",
    "    align=\"center\",\n",
    "    baseline=\"bottom\",\n",
    "    fontSize=16,\n",
    "    fontWeight=400,\n",
    "    angle=270,\n",
    "    color='black'\n",
    ").encode(\n",
    "    x=alt.value(10),  # pixels from left\n",
    "    y=alt.value(160),  # pixels from top\n",
    "    text=alt.value(['Rate of Adaptation'])\n",
    ")\n",
    "\n",
    "#y-axis label for both plots\n",
    "text2 = alt.Chart().mark_text(\n",
    "    align=\"center\",\n",
    "    baseline=\"bottom\",\n",
    "    fontSize=12,\n",
    "    fontWeight=400,\n",
    "    angle=270,\n",
    "    color='black'\n",
    ").encode(\n",
    "    x=alt.value(0),  # pixels from left\n",
    "    y=alt.value(160),  # pixels from top\n",
    "    text=alt.value(['Adaptive Mutations per Codon per Year (× 10⁻³)'])\n",
    ")\n",
    "\n",
    "#layer the points and error bars\n",
    "top = alt.layer(points_p, errorbars_p, vert_divider, data=df_polymerase)\n",
    "bottom = alt.layer(points_r, errorbars_r, vert_divider, enveloped_text, nonenveloped_text, data=df_receptorbinding)\n",
    "\n",
    "#stack the genes\n",
    "plot_layout = alt.vconcat(top, bottom, spacing=10)\n",
    "\n",
    "#add the yaxis label\n",
    "chart = alt.hconcat(text, text2, plot_layout, spacing=0).configure_axis(\n",
    "    grid=False\n",
    ").configure_view(\n",
    "    strokeWidth=0\n",
    ").configure_title(\n",
    "    fontSize=18,\n",
    "    anchor='start',\n",
    "    color='#c6c6c6'\n",
    ").resolve_scale(\n",
    "    y='independent',\n",
    "    x='independent',\n",
    ").configure_axisX(labelAngle=270)\n",
    "\n",
    "chart.save('../atlas-of-viral-adaptation/assets/adaptation_percodon_overview_plot.html')\n",
    "\n",
    "chart\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8af22995",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-5016fe9f268648658de371b8951ac176\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-5016fe9f268648658de371b8951ac176\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-5016fe9f268648658de371b8951ac176\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300, \"strokeWidth\": 0}, \"axis\": {\"grid\": false}, \"axisX\": {\"labelAngle\": 270}, \"title\": {\"anchor\": \"start\", \"color\": \"#c6c6c6\", \"fontSize\": 18}}, \"hconcat\": [{\"data\": {\"name\": \"empty\"}, \"mark\": {\"type\": \"text\", \"align\": \"center\", \"angle\": 270, \"baseline\": \"bottom\", \"color\": \"black\", \"fontSize\": 16, \"fontWeight\": 400}, \"encoding\": {\"text\": {\"value\": [\"Rate of Adaptation\"]}, \"x\": {\"value\": 10}, \"y\": {\"value\": 140}}}, {\"data\": {\"name\": \"empty\"}, \"mark\": {\"type\": \"text\", \"align\": \"center\", \"angle\": 270, \"baseline\": \"bottom\", \"color\": \"black\", \"fontSize\": 12, \"fontWeight\": 400}, \"encoding\": {\"text\": {\"value\": [\"Adaptive Mutations per Year\"]}, \"x\": {\"value\": 0}, \"y\": {\"value\": 140}}}, {\"vconcat\": [{\"layer\": [{\"data\": {\"name\": \"data-06a554f865048b667d6c6c7db9d5e2db\"}, \"mark\": {\"type\": \"point\", \"filled\": true, \"opacity\": 1, \"size\": 200}, \"encoding\": {\"color\": {\"field\": \"legible_name\", \"legend\": null, \"scale\": {\"domain\": [\"Influenza B/Vic\", \"229E\", \"Dengue 2-C\", \"Dengue 1-IV\", \"Influenza A/H3N2\", \"Lassa-A\", \"RSV-B\", \"RSV-A\", \"Dengue 3-II\", \"Dengue 4-I\", \"OC43-A\", \"Dengue 4-II\", \"Enterovirus D68\", \"Dengue 1-V\", \"Dengue 3-III\", \"Dengue 2-AI\", \"Influenza A/H1N1pdm\", \"Dengue 2-AA\", \"Mumps\", \"Lassa-B\", \"Parainfluenza-1\", \"Norovirus GII.4\", \"Parainfluenza-3\", \"NL63\", \"Influenza B/Yam\", \"Measles\", \"Dengue 1-III\", \"Dengue 1-I\"], \"range\": [\"#E06E85\", \"#76C7BE\", \"#2cccf2\", \"#0db0d6\", \"#9F2A44\", \"#9cc479\", \"#6C7593\", \"#0B194C\", \"#76def7\", \"#adebfa\", \"#208288\", \"#adebfa\", \"#a373a5\", \"#0db0d6\", \"#76def7\", \"#2cccf2\", \"#CC415A\", \"#2cccf2\", \"#eb520d\", \"#c4dcaf\", \"#f58451\", \"#d0b9d2\", \"#f79e76\", \"#BBE4D1\", \"#ED9AB0\", \"#F5AD52\", \"#0db0d6\", \"#0db0d6\"]}, \"type\": \"nominal\"}, \"href\": {\"field\": \"url\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"legible_name\", \"title\": \"Virus\", \"type\": \"nominal\"}, {\"field\": \"virus_family\", \"title\": \"Family\", \"type\": \"nominal\"}, {\"field\": \"genome_type\", \"title\": \"Genome Type\", \"type\": \"nominal\"}, {\"field\": \"enveloped\", \"title\": \"Enveloped\", \"type\": \"nominal\"}, {\"field\": \"legible_gene_name\", \"title\": \"Protein/Subunit\", \"type\": \"nominal\"}, {\"field\": \"legible_rate_pergene\", \"title\": \"Rate\", \"type\": \"nominal\"}], \"x\": {\"axis\": {\"labelExpr\": \"datum.label == 3 ? 'Measles' : datum.label == 6 ? 'Mumps' : datum.label == 9 ? 'Parainfluenza-1' : datum.label == 12 ? 'Parainfluenza-3' : datum.label == 15 ? 'Influenza A/H3N2' : datum.label == 18 ? 'Influenza A/H1N1pdm' : datum.label == 21 ? 'Influenza B/Vic' : datum.label == 24 ? 'Influenza B/Yam' : datum.label == 27 ? 'OC43-A' : datum.label == 30 ? '229E' : datum.label == 33 ? 'NL63' : datum.label == 36 ? 'RSV-A' : datum.label == 39 ? 'RSV-B' : datum.label == 42 ? 'Dengue 1-I' : datum.label == 43 ? 'Dengue 1-III' : datum.label == 44 ? 'Dengue 1-IV' : datum.label == 45 ? 'Dengue 1-V' : datum.label == 48 ? 'Dengue 2-AA' : datum.label == 49 ? 'Dengue 2-AI' : datum.label == 50 ? 'Dengue 2-C' : datum.label == 53 ? 'Dengue 3-II' : datum.label == 54 ? 'Dengue 3-III' : datum.label == 57 ? 'Dengue 4-I' : datum.label == 58 ? 'Dengue 4-II' : datum.label == 64 ? 'Enterovirus D68' : 'Norovirus GII.4'\", \"tickCount\": 26, \"values\": [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 43, 44, 45, 48, 49, 50, 53, 54, 57, 58, 64, 67]}, \"field\": \"xtick_pos\", \"title\": \"\", \"type\": \"quantitative\"}, \"y\": {\"axis\": {\"format\": \".1f\"}, \"field\": \"adaptive_muts_per_year\", \"title\": \"\", \"type\": \"quantitative\"}}, \"height\": 57.96833374243101, \"title\": \"Polymerase\", \"transform\": [{\"calculate\": \"('https://blab.github.io/atlas-of-viral-adaptation/' + datum.url_ending)\", \"as\": \"url\"}], \"width\": 800}, {\"data\": {\"name\": \"data-06a554f865048b667d6c6c7db9d5e2db\"}, \"mark\": \"errorbar\", \"encoding\": {\"color\": {\"field\": \"legible_name\", \"legend\": null, \"scale\": {\"domain\": [\"Influenza B/Vic\", \"229E\", \"Dengue 2-C\", \"Dengue 1-IV\", \"Influenza A/H3N2\", \"Lassa-A\", \"RSV-B\", \"RSV-A\", \"Dengue 3-II\", \"Dengue 4-I\", \"OC43-A\", \"Dengue 4-II\", \"Enterovirus D68\", \"Dengue 1-V\", \"Dengue 3-III\", \"Dengue 2-AI\", \"Influenza A/H1N1pdm\", \"Dengue 2-AA\", \"Mumps\", \"Lassa-B\", \"Parainfluenza-1\", \"Norovirus GII.4\", \"Parainfluenza-3\", \"NL63\", \"Influenza B/Yam\", \"Measles\", \"Dengue 1-III\", \"Dengue 1-I\"], \"range\": [\"#E06E85\", \"#76C7BE\", \"#2cccf2\", \"#0db0d6\", \"#9F2A44\", \"#9cc479\", \"#6C7593\", \"#0B194C\", \"#76def7\", \"#adebfa\", \"#208288\", \"#adebfa\", \"#a373a5\", \"#0db0d6\", \"#76def7\", \"#2cccf2\", \"#CC415A\", \"#2cccf2\", \"#eb520d\", \"#c4dcaf\", \"#f58451\", \"#d0b9d2\", \"#f79e76\", \"#BBE4D1\", \"#ED9AB0\", \"#F5AD52\", \"#0db0d6\", \"#0db0d6\"]}, \"type\": \"nominal\"}, \"tooltip\": {\"value\": null}, \"x\": {\"axis\": {\"labels\": false}, \"field\": \"xtick_pos\", \"sort\": [\"Measles\", \"Mumps\", \"Parainfluenza-1\", \"Parainfluenza-3\", \"Influenza A/H3N2\", \"Influenza A/H1N1pdm\", \"Influenza B/Vic\", \"Influenza B/Yam\", \"OC43-A\", \"229E\", \"NL63\", \"RSV-A\", \"RSV-B\", \"Dengue 1-I\", \"Dengue 1-III\", \"Dengue 1-IV\", \"Dengue 1-V\", \"Dengue 2-AA\", \"Dengue 2-AI\", \"Dengue 2-C\", \"Dengue 3-II\", \"Dengue 3-III\", \"Dengue 4-I\", \"Dengue 4-II\", \"Enterovirus D68\", \"Norovirus GII.4\"], \"title\": \"\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"lower_95ci_mutspergene\", \"title\": \"\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"upper_95ci_mutspergene\"}}, \"height\": 57.96833374243101, \"width\": 800}, {\"data\": {\"name\": \"data-0209ec4f059fa6853fca90f6e38626c5\"}, \"mark\": {\"type\": \"rule\", \"strokeDash\": [3, 5]}, \"encoding\": {\"x\": {\"field\": \"x\", \"type\": \"quantitative\"}}}], \"data\": {\"name\": \"data-06a554f865048b667d6c6c7db9d5e2db\"}}, {\"layer\": [{\"data\": {\"name\": \"data-e5bd9ca9abcbf571067ef56c644e2ec6\"}, \"mark\": {\"type\": \"point\", \"filled\": true, \"opacity\": 1, \"size\": 200}, \"encoding\": {\"color\": {\"field\": \"legible_name\", \"legend\": null, \"scale\": {\"domain\": [\"Influenza B/Vic\", \"229E\", \"Dengue 2-C\", \"Dengue 1-IV\", \"Influenza A/H3N2\", \"Lassa-A\", \"RSV-B\", \"RSV-A\", \"Dengue 3-II\", \"Dengue 4-I\", \"OC43-A\", \"Dengue 4-II\", \"Enterovirus D68\", \"Dengue 1-V\", \"Dengue 3-III\", \"Dengue 2-AI\", \"Influenza A/H1N1pdm\", \"Dengue 2-AA\", \"Mumps\", \"Lassa-B\", \"Parainfluenza-1\", \"Norovirus GII.4\", \"Parainfluenza-3\", \"NL63\", \"Influenza B/Yam\", \"Measles\", \"Dengue 1-III\", \"Dengue 1-I\"], \"range\": [\"#E06E85\", \"#76C7BE\", \"#2cccf2\", \"#0db0d6\", \"#9F2A44\", \"#9cc479\", \"#6C7593\", \"#0B194C\", \"#76def7\", \"#adebfa\", \"#208288\", \"#adebfa\", \"#a373a5\", \"#0db0d6\", \"#76def7\", \"#2cccf2\", \"#CC415A\", \"#2cccf2\", \"#eb520d\", \"#c4dcaf\", \"#f58451\", \"#d0b9d2\", \"#f79e76\", \"#BBE4D1\", \"#ED9AB0\", \"#F5AD52\", \"#0db0d6\", \"#0db0d6\"]}, \"type\": \"nominal\"}, \"href\": {\"field\": \"url\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"legible_name\", \"title\": \"Virus\", \"type\": \"nominal\"}, {\"field\": \"virus_family\", \"title\": \"Family\", \"type\": \"nominal\"}, {\"field\": \"genome_type\", \"title\": \"Genome Type\", \"type\": \"nominal\"}, {\"field\": \"enveloped\", \"title\": \"Enveloped\", \"type\": \"nominal\"}, {\"field\": \"legible_gene_name\", \"title\": \"Protein/Subunit\", \"type\": \"nominal\"}, {\"field\": \"legible_rate_pergene\", \"title\": \"Rate\", \"type\": \"nominal\"}], \"x\": {\"axis\": {\"labelExpr\": \"datum.label == 3 ? 'Measles' : datum.label == 6 ? 'Mumps' : datum.label == 9 ? 'Parainfluenza-1' : datum.label == 12 ? 'Parainfluenza-3' : datum.label == 15 ? 'Influenza A/H3N2' : datum.label == 18 ? 'Influenza A/H1N1pdm' : datum.label == 21 ? 'Influenza B/Vic' : datum.label == 24 ? 'Influenza B/Yam' : datum.label == 27 ? 'OC43-A' : datum.label == 30 ? '229E' : datum.label == 33 ? 'NL63' : datum.label == 36 ? 'RSV-A' : datum.label == 39 ? 'RSV-B' : datum.label == 42 ? 'Dengue 1-I' : datum.label == 43 ? 'Dengue 1-III' : datum.label == 44 ? 'Dengue 1-IV' : datum.label == 45 ? 'Dengue 1-V' : datum.label == 48 ? 'Dengue 2-AA' : datum.label == 49 ? 'Dengue 2-AI' : datum.label == 50 ? 'Dengue 2-C' : datum.label == 53 ? 'Dengue 3-II' : datum.label == 54 ? 'Dengue 3-III' : datum.label == 57 ? 'Dengue 4-I' : datum.label == 58 ? 'Dengue 4-II' : datum.label == 64 ? 'Enterovirus D68' : 'Norovirus GII.4'\", \"tickCount\": 26, \"values\": [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 43, 44, 45, 48, 49, 50, 53, 54, 57, 58, 64, 67]}, \"field\": \"xtick_pos\", \"title\": \"\", \"type\": \"quantitative\"}, \"y\": {\"axis\": {\"format\": \".1f\"}, \"field\": \"adaptive_muts_per_year\", \"title\": \"\", \"type\": \"quantitative\"}}, \"height\": 250, \"title\": \"Receptor-Binding\", \"transform\": [{\"calculate\": \"('https://blab.github.io/atlas-of-viral-adaptation/' + datum.url_ending)\", \"as\": \"url\"}], \"width\": 800}, {\"data\": {\"name\": \"data-e5bd9ca9abcbf571067ef56c644e2ec6\"}, \"mark\": \"errorbar\", \"encoding\": {\"color\": {\"field\": \"legible_name\", \"legend\": null, \"scale\": {\"domain\": [\"Influenza B/Vic\", \"229E\", \"Dengue 2-C\", \"Dengue 1-IV\", \"Influenza A/H3N2\", \"Lassa-A\", \"RSV-B\", \"RSV-A\", \"Dengue 3-II\", \"Dengue 4-I\", \"OC43-A\", \"Dengue 4-II\", \"Enterovirus D68\", \"Dengue 1-V\", \"Dengue 3-III\", \"Dengue 2-AI\", \"Influenza A/H1N1pdm\", \"Dengue 2-AA\", \"Mumps\", \"Lassa-B\", \"Parainfluenza-1\", \"Norovirus GII.4\", \"Parainfluenza-3\", \"NL63\", \"Influenza B/Yam\", \"Measles\", \"Dengue 1-III\", \"Dengue 1-I\"], \"range\": [\"#E06E85\", \"#76C7BE\", \"#2cccf2\", \"#0db0d6\", \"#9F2A44\", \"#9cc479\", \"#6C7593\", \"#0B194C\", \"#76def7\", \"#adebfa\", \"#208288\", \"#adebfa\", \"#a373a5\", \"#0db0d6\", \"#76def7\", \"#2cccf2\", \"#CC415A\", \"#2cccf2\", \"#eb520d\", \"#c4dcaf\", \"#f58451\", \"#d0b9d2\", \"#f79e76\", \"#BBE4D1\", \"#ED9AB0\", \"#F5AD52\", \"#0db0d6\", \"#0db0d6\"]}, \"type\": \"nominal\"}, \"tooltip\": {\"value\": null}, \"x\": {\"field\": \"xtick_pos\", \"sort\": [\"Measles\", \"Mumps\", \"Parainfluenza-1\", \"Parainfluenza-3\", \"Influenza A/H3N2\", \"Influenza A/H1N1pdm\", \"Influenza B/Vic\", \"Influenza B/Yam\", \"OC43-A\", \"229E\", \"NL63\", \"RSV-A\", \"RSV-B\", \"Dengue 1-I\", \"Dengue 1-III\", \"Dengue 1-IV\", \"Dengue 1-V\", \"Dengue 2-AA\", \"Dengue 2-AI\", \"Dengue 2-C\", \"Dengue 3-II\", \"Dengue 3-III\", \"Dengue 4-I\", \"Dengue 4-II\", \"Enterovirus D68\", \"Norovirus GII.4\"], \"title\": \"\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"lower_95ci_mutspergene\", \"title\": \"\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"upper_95ci_mutspergene\"}}, \"height\": 250, \"width\": 800}, {\"data\": {\"name\": \"data-0209ec4f059fa6853fca90f6e38626c5\"}, \"mark\": {\"type\": \"rule\", \"strokeDash\": [3, 5]}, \"encoding\": {\"x\": {\"field\": \"x\", \"type\": \"quantitative\"}}}, {\"data\": {\"name\": \"data-0209ec4f059fa6853fca90f6e38626c5\"}, \"mark\": {\"type\": \"text\", \"align\": \"right\", \"color\": \"#6e6e6e\", \"dx\": -10, \"dy\": -120}, \"encoding\": {\"text\": {\"value\": [\"enveloped\", \"viruses\"]}, \"x\": {\"field\": \"x\", \"type\": \"quantitative\"}}}, {\"data\": {\"name\": \"data-0209ec4f059fa6853fca90f6e38626c5\"}, \"mark\": {\"type\": \"text\", \"align\": \"left\", \"color\": \"#6e6e6e\", \"dx\": 10, \"dy\": -120}, \"encoding\": {\"text\": {\"value\": [\"non-enveloped\", \"viruses\"]}, \"x\": {\"field\": \"x\", \"type\": \"quantitative\"}}}], \"data\": {\"name\": \"data-e5bd9ca9abcbf571067ef56c644e2ec6\"}}], \"spacing\": 10}], \"resolve\": {\"scale\": {\"x\": \"independent\", \"y\": \"independent\"}}, \"spacing\": 0, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"empty\": [{}], \"data-06a554f865048b667d6c6c7db9d5e2db\": [{\"virus\": \"measles\", \"subtype\": null, \"virus_and_subtype\": \"measles\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0696477489789021, \"lower_95ci\": 0.0019629917949619, \"upper_95ci\": 0.1477218174503109, \"ci\": \"[0.00196299179496196, 0.14772181745031096]\", \"len_in_codons\": 2184, \"adaptive_muts_per_year\": 0.1521106837699223, \"lower_95ci_mutspergene\": 0.0042871740801969, \"upper_95ci_mutspergene\": 0.3226244493114791, \"legible_name\": \"Measles\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#F5AD52\", \"index\": 0, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"measles\", \"xtick_pos\": 3, \"legible_rate_percodon\": \"0.07\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.15 muts per year\"}, {\"virus\": \"mumps\", \"subtype\": null, \"virus_and_subtype\": \"mumps\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": -0.0036394625647537, \"lower_95ci\": -0.0116868160093167, \"upper_95ci\": 0.016236636926292, \"ci\": \"[-0.011686816009316748, 0.016236636926292045]\", \"len_in_codons\": 2262, \"adaptive_muts_per_year\": -0.008232464321473, \"lower_95ci_mutspergene\": -0.0264355778130744, \"upper_95ci_mutspergene\": 0.0367272727272726, \"legible_name\": \"Mumps\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#eb520d\", \"index\": 3, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"mumps\", \"xtick_pos\": 6, \"legible_rate_percodon\": \"-0.0\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"-0.01 muts per year\"}, {\"virus\": \"parainfluenza\", \"subtype\": \"1\", \"virus_and_subtype\": \"parainfluenza_1\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1245544489235666, \"lower_95ci\": 0.0213585194689124, \"upper_95ci\": 0.2496937595953217, \"ci\": \"[0.021358519468912436, 0.24969375959532172]\", \"len_in_codons\": 2224, \"adaptive_muts_per_year\": 0.2770090944060122, \"lower_95ci_mutspergene\": 0.0475013472988612, \"upper_95ci_mutspergene\": 0.5553189213399955, \"legible_name\": \"Parainfluenza-1\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#f58451\", \"index\": 21, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"parainfluenza_1\", \"xtick_pos\": 9, \"legible_rate_percodon\": \"0.12\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.28 muts per year\"}, {\"virus\": \"parainfluenza\", \"subtype\": \"3\", \"virus_and_subtype\": \"parainfluenza_3\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0168440508390062, \"lower_95ci\": 0.0, \"upper_95ci\": 0.0518289545156088, \"ci\": \"[0.0, 0.05182895451560882]\", \"len_in_codons\": 2234, \"adaptive_muts_per_year\": 0.0376296095743399, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1157858843878701, \"legible_name\": \"Parainfluenza-3\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#f79e76\", \"index\": 21, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"parainfluenza_3\", \"xtick_pos\": 12, \"legible_rate_percodon\": \"0.02\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.04 muts per year\"}, {\"virus\": \"h3n2\", \"subtype\": null, \"virus_and_subtype\": \"h3n2\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.3804195219421517, \"lower_95ci\": 0.2019456795663166, \"upper_95ci\": 0.587181685626881, \"ci\": \"[0.2019456795663166, 0.587181685626881]\", \"len_in_codons\": 758, \"adaptive_muts_per_year\": 0.288357997632151, \"lower_95ci_mutspergene\": 0.153074825111268, \"upper_95ci_mutspergene\": 0.4450837177051758, \"legible_name\": \"Influenza A/H3N2\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#9F2A44\", \"index\": 24, \"legible_gene_name\": \"PB1 (Polymerase)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"h3n2\", \"xtick_pos\": 15, \"legible_rate_percodon\": \"0.38\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.29 muts per year\"}, {\"virus\": \"h1n1pdm\", \"subtype\": null, \"virus_and_subtype\": \"h1n1pdm\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2895886158625782, \"lower_95ci\": -0.001560244403758, \"upper_95ci\": 1.132737866649248, \"ci\": \"[-0.0015602444037580434, 1.132737866649248]\", \"len_in_codons\": 758, \"adaptive_muts_per_year\": 0.2195081708238342, \"lower_95ci_mutspergene\": -0.0011826652580485, \"upper_95ci_mutspergene\": 0.85861530292013, \"legible_name\": \"Influenza A/H1N1pdm\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#CC415A\", \"index\": 6, \"legible_gene_name\": \"PB1 (Polymerase)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"h1n1pdm\", \"xtick_pos\": 18, \"legible_rate_percodon\": \"0.29\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.22 muts per year\"}, {\"virus\": \"vic\", \"subtype\": null, \"virus_and_subtype\": \"vic\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1515692519238777, \"lower_95ci\": 0.0023655998857976, \"upper_95ci\": 0.315418668570784, \"ci\": \"[0.0023655998857976504, 0.31541866857078404]\", \"len_in_codons\": 753, \"adaptive_muts_per_year\": 0.1141316466986799, \"lower_95ci_mutspergene\": 0.0017812967140056, \"upper_95ci_mutspergene\": 0.2375102574338004, \"legible_name\": \"Influenza B/Vic\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#E06E85\", \"index\": 15, \"legible_gene_name\": \"PB1 (Polymerase)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"vic\", \"xtick_pos\": 21, \"legible_rate_percodon\": \"0.15\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"yam\", \"subtype\": null, \"virus_and_subtype\": \"yam\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1049109846029593, \"lower_95ci\": -0.0070505300278014, \"upper_95ci\": 0.2730279027767181, \"ci\": \"[-0.00705053002780142, 0.2730279027767181]\", \"len_in_codons\": 753, \"adaptive_muts_per_year\": 0.0789979714060283, \"lower_95ci_mutspergene\": -0.0053090491109344, \"upper_95ci_mutspergene\": 0.2055900107908687, \"legible_name\": \"Influenza B/Yam\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#ED9AB0\", \"index\": 9, \"legible_gene_name\": \"PB1 (Polymerase)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"yam\", \"xtick_pos\": 24, \"legible_rate_percodon\": \"0.1\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.08 muts per year\"}, {\"virus\": \"oc43\", \"subtype\": \"a\", \"virus_and_subtype\": \"oc43_a\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0368813805325201, \"lower_95ci\": 0.0, \"upper_95ci\": 0.1475255221300805, \"ci\": \"[0.0, 0.14752552213008055]\", \"len_in_codons\": 928, \"adaptive_muts_per_year\": 0.0342259211341786, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1369036845367147, \"legible_name\": \"OC43-A\", \"virus_family\": \"Coronavirus\", \"color\": \"#208288\", \"index\": 36, \"legible_gene_name\": \"RDRP (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"oc43_a\", \"xtick_pos\": 27, \"legible_rate_percodon\": \"0.04\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.03 muts per year\"}, {\"virus\": \"229e\", \"subtype\": null, \"virus_and_subtype\": \"229e\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0369532263703259, \"lower_95ci\": 0.0, \"upper_95ci\": 0.1349980102763556, \"ci\": \"[0.0, 0.13499801027635566]\", \"len_in_codons\": 927, \"adaptive_muts_per_year\": 0.0342556408452921, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1251431555261817, \"legible_name\": \"229E\", \"virus_family\": \"Coronavirus\", \"color\": \"#76C7BE\", \"index\": 18, \"legible_gene_name\": \"RDRP (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"229e\", \"xtick_pos\": 30, \"legible_rate_percodon\": \"0.04\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.03 muts per year\"}, {\"virus\": \"nl63\", \"subtype\": null, \"virus_and_subtype\": \"nl63\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0, \"lower_95ci\": 0.0, \"upper_95ci\": 0.0, \"ci\": \"[0.0, 0.0]\", \"len_in_codons\": 927, \"adaptive_muts_per_year\": 0.0, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.0, \"legible_name\": \"NL63\", \"virus_family\": \"Coronavirus\", \"color\": \"#BBE4D1\", \"index\": 27, \"legible_gene_name\": \"RDRP (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"nl63\", \"xtick_pos\": 33, \"legible_rate_percodon\": \"0.0\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.0 muts per year\"}, {\"virus\": \"rsv\", \"subtype\": \"A\", \"virus_and_subtype\": \"rsv_A\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1791526081703916, \"lower_95ci\": 0.0734714909962797, \"upper_95ci\": 0.2786727160334553, \"ci\": \"[0.07347149099627973, 0.27867271603345534]\", \"len_in_codons\": 2166, \"adaptive_muts_per_year\": 0.3880445492970683, \"lower_95ci_mutspergene\": 0.1591392494979418, \"upper_95ci_mutspergene\": 0.6036051029284643, \"legible_name\": \"RSV-A\", \"virus_family\": \"Pneumovirus\", \"color\": \"#0B194C\", \"index\": 39, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"rsv_a\", \"xtick_pos\": 36, \"legible_rate_percodon\": \"0.18\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.39 muts per year\"}, {\"virus\": \"rsv\", \"subtype\": \"B\", \"virus_and_subtype\": \"rsv_B\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.227456322682962, \"lower_95ci\": 0.0999231081717831, \"upper_95ci\": 0.325143188036216, \"ci\": \"[0.09992310817178313, 0.32514318803621606]\", \"len_in_codons\": 2166, \"adaptive_muts_per_year\": 0.4926703949312957, \"lower_95ci_mutspergene\": 0.2164334523000822, \"upper_95ci_mutspergene\": 0.704260145286444, \"legible_name\": \"RSV-B\", \"virus_family\": \"Pneumovirus\", \"color\": \"#6C7593\", \"index\": 39, \"legible_gene_name\": \"L (Polymerase)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"rsv_b\", \"xtick_pos\": 39, \"legible_rate_percodon\": \"0.23\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.49 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_I\", \"virus_and_subtype\": \"dengue_denv1_I\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2548158661827924, \"lower_95ci\": 0.0359724661139091, \"upper_95ci\": 0.5927314906874545, \"ci\": \"[0.03597246611390916, 0.5927314906874545]\", \"len_in_codons\": 899, \"adaptive_muts_per_year\": 0.2290794636983303, \"lower_95ci_mutspergene\": 0.0323392470364043, \"upper_95ci_mutspergene\": 0.5328656101280216, \"legible_name\": \"Dengue 1-I\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_i\", \"xtick_pos\": 42, \"legible_rate_percodon\": \"0.25\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.23 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_III\", \"virus_and_subtype\": \"dengue_denv1_III\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0557099382380511, \"lower_95ci\": 0.0, \"upper_95ci\": 0.1650902814782644, \"ci\": \"[0.0, 0.16509028147826446]\", \"len_in_codons\": 899, \"adaptive_muts_per_year\": 0.050083234476008, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1484161630489597, \"legible_name\": \"Dengue 1-III\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_iii\", \"xtick_pos\": 43, \"legible_rate_percodon\": \"0.06\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.05 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_IV\", \"virus_and_subtype\": \"dengue_denv1_IV\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1536898027582185, \"lower_95ci\": 0.0093388951033292, \"upper_95ci\": 0.3542488014812919, \"ci\": \"[0.009338895103329251, 0.3542488014812919]\", \"len_in_codons\": 899, \"adaptive_muts_per_year\": 0.1381671326796384, \"lower_95ci_mutspergene\": 0.0083956666978929, \"upper_95ci_mutspergene\": 0.3184696725316814, \"legible_name\": \"Dengue 1-IV\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_iv\", \"xtick_pos\": 44, \"legible_rate_percodon\": \"0.15\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.14 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_V\", \"virus_and_subtype\": \"dengue_denv1_V\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1207176423891461, \"lower_95ci\": 0.0, \"upper_95ci\": 0.2506198310284858, \"ci\": \"[0.0, 0.2506198310284858]\", \"len_in_codons\": 899, \"adaptive_muts_per_year\": 0.1085251605078424, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.2253072280946087, \"legible_name\": \"Dengue 1-V\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_v\", \"xtick_pos\": 45, \"legible_rate_percodon\": \"0.12\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_AA\", \"virus_and_subtype\": \"dengue_denv2_AA\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2300077187985516, \"lower_95ci\": 0.0820793317286246, \"upper_95ci\": 0.4116735028777324, \"ci\": \"[0.08207933172862468, 0.41167350287773247]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.2070069469186964, \"lower_95ci_mutspergene\": 0.0738713985557622, \"upper_95ci_mutspergene\": 0.3705061525899592, \"legible_name\": \"Dengue 2-AA\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_aa\", \"xtick_pos\": 48, \"legible_rate_percodon\": \"0.23\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.21 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_AI\", \"virus_and_subtype\": \"dengue_denv2_AI\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1557132167090561, \"lower_95ci\": 0.0107601070156147, \"upper_95ci\": 0.3535455238972207, \"ci\": \"[0.010760107015614785, 0.35354552389722077]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.1401418950381505, \"lower_95ci_mutspergene\": 0.0096840963140533, \"upper_95ci_mutspergene\": 0.3181909715074987, \"legible_name\": \"Dengue 2-AI\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_ai\", \"xtick_pos\": 49, \"legible_rate_percodon\": \"0.16\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.14 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_C\", \"virus_and_subtype\": \"dengue_denv2_C\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": -0.0617916284955231, \"lower_95ci\": -0.1282134695879142, \"upper_95ci\": -0.0126741769450911, \"ci\": \"[-0.1282134695879142, -0.012674176945091181]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": -0.0556124656459708, \"lower_95ci_mutspergene\": -0.1153921226291227, \"upper_95ci_mutspergene\": -0.011406759250582, \"legible_name\": \"Dengue 2-C\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_c\", \"xtick_pos\": 50, \"legible_rate_percodon\": \"-0.06\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"-0.06 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv3_II\", \"virus_and_subtype\": \"dengue_denv3_II\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1270526611523327, \"lower_95ci\": 0.045667845029431, \"upper_95ci\": 0.204669909644794, \"ci\": \"[0.04566784502943106, 0.20466990964479406]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.1143473950370994, \"lower_95ci_mutspergene\": 0.0411010605264879, \"upper_95ci_mutspergene\": 0.1842029186803146, \"legible_name\": \"Dengue 3-II\", \"virus_family\": \"Flavivirus\", \"color\": \"#76def7\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv3_ii\", \"xtick_pos\": 53, \"legible_rate_percodon\": \"0.13\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv3_III\", \"virus_and_subtype\": \"dengue_denv3_III\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.1179126365521947, \"lower_95ci\": 0.0204133578039758, \"upper_95ci\": 0.248946148912938, \"ci\": \"[0.02041335780397588, 0.24894614891293806]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.1061213728969753, \"lower_95ci_mutspergene\": 0.0183720220235782, \"upper_95ci_mutspergene\": 0.2240515340216442, \"legible_name\": \"Dengue 3-III\", \"virus_family\": \"Flavivirus\", \"color\": \"#76def7\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv3_iii\", \"xtick_pos\": 54, \"legible_rate_percodon\": \"0.12\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv4_I\", \"virus_and_subtype\": \"dengue_denv4_I\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2104987551233288, \"lower_95ci\": 0.0889963151806821, \"upper_95ci\": 0.3305517123186876, \"ci\": \"[0.08899631518068213, 0.3305517123186876]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.189448879610996, \"lower_95ci_mutspergene\": 0.0800966836626139, \"upper_95ci_mutspergene\": 0.2974965410868188, \"legible_name\": \"Dengue 4-I\", \"virus_family\": \"Flavivirus\", \"color\": \"#adebfa\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv4_i\", \"xtick_pos\": 57, \"legible_rate_percodon\": \"0.21\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.19 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv4_II\", \"virus_and_subtype\": \"dengue_denv4_II\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.0723489159160415, \"lower_95ci\": 0.0286215625878451, \"upper_95ci\": 0.1252053439161044, \"ci\": \"[0.028621562587845113, 0.1252053439161044]\", \"len_in_codons\": 900, \"adaptive_muts_per_year\": 0.0651140243244374, \"lower_95ci_mutspergene\": 0.0257594063290606, \"upper_95ci_mutspergene\": 0.1126848095244939, \"legible_name\": \"Dengue 4-II\", \"virus_family\": \"Flavivirus\", \"color\": \"#adebfa\", \"index\": 33, \"legible_gene_name\": \"NS5 (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv4_ii\", \"xtick_pos\": 58, \"legible_rate_percodon\": \"0.07\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.07 muts per year\"}, {\"virus\": \"enterovirusd68\", \"subtype\": null, \"virus_and_subtype\": \"enterovirusd68\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.3421578702666206, \"lower_95ci\": 0.0312003072602813, \"upper_95ci\": 0.735949789390216, \"ci\": \"[0.031200307260281307, 0.735949789390216]\", \"len_in_codons\": 457, \"adaptive_muts_per_year\": 0.1563661467118456, \"lower_95ci_mutspergene\": 0.0142585404179485, \"upper_95ci_mutspergene\": 0.3363290537513287, \"legible_name\": \"Enterovirus D68\", \"virus_family\": \"Picornavirus\", \"color\": \"#a373a5\", \"index\": 30, \"legible_gene_name\": \"3D (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"False\", \"url_ending\": \"enterovirusd68\", \"xtick_pos\": 64, \"legible_rate_percodon\": \"0.34\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.16 muts per year\"}, {\"virus\": \"norovirus\", \"subtype\": null, \"virus_and_subtype\": \"norovirus\", \"gene\": \"polymerase\", \"adaptive_subs_per_codon_per_year\": 0.2909630611667262, \"lower_95ci\": 0.0404426097320981, \"upper_95ci\": 0.5874570070327857, \"ci\": \"[0.040442609732098164, 0.5874570070327857]\", \"len_in_codons\": 510, \"adaptive_muts_per_year\": 0.1483911611950304, \"lower_95ci_mutspergene\": 0.02062573096337, \"upper_95ci_mutspergene\": 0.2996030735867207, \"legible_name\": \"Norovirus GII.4\", \"virus_family\": \"Caliciviridae\", \"color\": \"#d0b9d2\", \"index\": 12, \"legible_gene_name\": \"RDRP (Polymerase)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"False\", \"url_ending\": \"norovirus_gii4\", \"xtick_pos\": 67, \"legible_rate_percodon\": \"0.29\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.15 muts per year\"}], \"data-0209ec4f059fa6853fca90f6e38626c5\": [{\"x\": 61}], \"data-e5bd9ca9abcbf571067ef56c644e2ec6\": [{\"virus\": \"measles\", \"subtype\": null, \"virus_and_subtype\": \"measles\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.1488760781105891, \"lower_95ci\": 0.0, \"upper_95ci\": 0.464536965766201, \"ci\": \"[0.0, 0.46453696576620107]\", \"len_in_codons\": 618, \"adaptive_muts_per_year\": 0.0920054162723441, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.2870838448435123, \"legible_name\": \"Measles\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#F5AD52\", \"index\": 2, \"legible_gene_name\": \"H (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"measles\", \"xtick_pos\": 3, \"legible_rate_percodon\": \"0.15\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.09 muts per year\"}, {\"virus\": \"mumps\", \"subtype\": null, \"virus_and_subtype\": \"mumps\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.0177121918434842, \"lower_95ci\": 0.0, \"upper_95ci\": 0.0837561442164208, \"ci\": \"[0.0, 0.08375614421642087]\", \"len_in_codons\": 583, \"adaptive_muts_per_year\": 0.0103262078447513, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.0488298320781733, \"legible_name\": \"Mumps\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#eb520d\", \"index\": 5, \"legible_gene_name\": \"HN (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"mumps\", \"xtick_pos\": 6, \"legible_rate_percodon\": \"0.02\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.01 muts per year\"}, {\"virus\": \"parainfluenza\", \"subtype\": \"1\", \"virus_and_subtype\": \"parainfluenza_1\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.5301824829604939, \"lower_95ci\": -0.0807613335513118, \"upper_95ci\": 1.0682467463132916, \"ci\": \"[-0.08076133355131188, 1.0682467463132916]\", \"len_in_codons\": 576, \"adaptive_muts_per_year\": 0.3053851101852444, \"lower_95ci_mutspergene\": -0.0465185281255556, \"upper_95ci_mutspergene\": 0.6153101258764561, \"legible_name\": \"Parainfluenza-1\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#f58451\", \"index\": 23, \"legible_gene_name\": \"HN (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"parainfluenza_1\", \"xtick_pos\": 9, \"legible_rate_percodon\": \"0.53\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.31 muts per year\"}, {\"virus\": \"parainfluenza\", \"subtype\": \"3\", \"virus_and_subtype\": \"parainfluenza_3\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.0, \"lower_95ci\": 0.0, \"upper_95ci\": 0.0, \"ci\": \"[0.0, 0.0]\", \"len_in_codons\": 573, \"adaptive_muts_per_year\": 0.0, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.0, \"legible_name\": \"Parainfluenza-3\", \"virus_family\": \"Paramyxovirus\", \"color\": \"#f79e76\", \"index\": 23, \"legible_gene_name\": \"HN (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"parainfluenza_3\", \"xtick_pos\": 12, \"legible_rate_percodon\": \"0.0\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.0 muts per year\"}, {\"virus\": \"h3n2\", \"subtype\": null, \"virus_and_subtype\": \"h3n2\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 3.968017826307325, \"lower_95ci\": 2.958853820205344, \"upper_95ci\": 4.99468539503025, \"ci\": \"[2.9588538202053436, 4.99468539503025]\", \"len_in_codons\": 329, \"adaptive_muts_per_year\": 1.3054778648551095, \"lower_95ci_mutspergene\": 0.973462906847558, \"upper_95ci_mutspergene\": 1.6432514949649522, \"legible_name\": \"Influenza A/H3N2\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#9F2A44\", \"index\": 26, \"legible_gene_name\": \"HA1 (Receptor Binding)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"h3n2\", \"xtick_pos\": 15, \"legible_rate_percodon\": \"3.97\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"1.31 muts per year\"}, {\"virus\": \"h1n1pdm\", \"subtype\": null, \"virus_and_subtype\": \"h1n1pdm\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 3.081395153594903, \"lower_95ci\": 1.3370908325016322, \"upper_95ci\": 5.264504569755848, \"ci\": \"[1.3370908325016322, 5.264504569755848]\", \"len_in_codons\": 327, \"adaptive_muts_per_year\": 1.0076162152255332, \"lower_95ci_mutspergene\": 0.4372287022280337, \"upper_95ci_mutspergene\": 1.7214929943101624, \"legible_name\": \"Influenza A/H1N1pdm\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#CC415A\", \"index\": 8, \"legible_gene_name\": \"HA1 (Receptor Binding)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"h1n1pdm\", \"xtick_pos\": 18, \"legible_rate_percodon\": \"3.08\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"1.01 muts per year\"}, {\"virus\": \"vic\", \"subtype\": null, \"virus_and_subtype\": \"vic\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.1534834193373809, \"lower_95ci\": 0.5371777048905576, \"upper_95ci\": 1.7054541798505731, \"ci\": \"[0.5371777048905576, 1.7054541798505731]\", \"len_in_codons\": 346, \"adaptive_muts_per_year\": 0.3991052630907338, \"lower_95ci_mutspergene\": 0.1858634858921329, \"upper_95ci_mutspergene\": 0.5900871462282983, \"legible_name\": \"Influenza B/Vic\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#E06E85\", \"index\": 17, \"legible_gene_name\": \"HA1 (Receptor Binding)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"vic\", \"xtick_pos\": 21, \"legible_rate_percodon\": \"1.15\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.4 muts per year\"}, {\"virus\": \"yam\", \"subtype\": null, \"virus_and_subtype\": \"yam\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.5005224136436983, \"lower_95ci\": 0.5579713024604996, \"upper_95ci\": 2.221057553942913, \"ci\": \"[0.5579713024604996, 2.221057553942913]\", \"len_in_codons\": 346, \"adaptive_muts_per_year\": 0.5191807551207196, \"lower_95ci_mutspergene\": 0.1930580706513328, \"upper_95ci_mutspergene\": 0.7684859136642479, \"legible_name\": \"Influenza B/Yam\", \"virus_family\": \"Orthomyxovirus\", \"color\": \"#ED9AB0\", \"index\": 11, \"legible_gene_name\": \"HA1 (Receptor Binding)\", \"genome_type\": \"Negative-sense, segmented\", \"enveloped\": \"True\", \"url_ending\": \"yam\", \"xtick_pos\": 24, \"legible_rate_percodon\": \"1.5\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.52 muts per year\"}, {\"virus\": \"oc43\", \"subtype\": \"a\", \"virus_and_subtype\": \"oc43_a\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.866573224980902, \"lower_95ci\": 1.0689984441339817, \"upper_95ci\": 2.828605874053624, \"ci\": \"[1.0689984441339815, 2.828605874053624]\", \"len_in_codons\": 753, \"adaptive_muts_per_year\": 1.405529638410619, \"lower_95ci_mutspergene\": 0.804955828432888, \"upper_95ci_mutspergene\": 2.129940223162379, \"legible_name\": \"OC43-A\", \"virus_family\": \"Coronavirus\", \"color\": \"#208288\", \"index\": 38, \"legible_gene_name\": \"S1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"oc43_a\", \"xtick_pos\": 27, \"legible_rate_percodon\": \"1.87\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"1.41 muts per year\"}, {\"virus\": \"229e\", \"subtype\": null, \"virus_and_subtype\": \"229e\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.4851663127760144, \"lower_95ci\": 0.9834867527798624, \"upper_95ci\": 2.068365846820297, \"ci\": \"[0.9834867527798623, 2.0683658468202974]\", \"len_in_codons\": 545, \"adaptive_muts_per_year\": 0.8094156404629278, \"lower_95ci_mutspergene\": 0.5360002802650249, \"upper_95ci_mutspergene\": 1.127259386517062, \"legible_name\": \"229E\", \"virus_family\": \"Coronavirus\", \"color\": \"#76C7BE\", \"index\": 20, \"legible_gene_name\": \"S1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"229e\", \"xtick_pos\": 30, \"legible_rate_percodon\": \"1.49\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.81 muts per year\"}, {\"virus\": \"nl63\", \"subtype\": null, \"virus_and_subtype\": \"nl63\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.3233562876592263, \"lower_95ci\": 0.1399656586773711, \"upper_95ci\": 0.5437555140139991, \"ci\": \"[0.13996565867737112, 0.5437555140139991]\", \"len_in_codons\": 734, \"adaptive_muts_per_year\": 0.2373435151418721, \"lower_95ci_mutspergene\": 0.1027347934691904, \"upper_95ci_mutspergene\": 0.3991165472862754, \"legible_name\": \"NL63\", \"virus_family\": \"Coronavirus\", \"color\": \"#BBE4D1\", \"index\": 29, \"legible_gene_name\": \"S1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"nl63\", \"xtick_pos\": 33, \"legible_rate_percodon\": \"0.32\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.24 muts per year\"}, {\"virus\": \"rsv\", \"subtype\": \"A\", \"virus_and_subtype\": \"rsv_A\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.9300665866808224, \"lower_95ci\": 1.1735832138549698, \"upper_95ci\": 2.878035332770765, \"ci\": \"[1.1735832138549698, 2.878035332770765]\", \"len_in_codons\": 298, \"adaptive_muts_per_year\": 0.5751598428308852, \"lower_95ci_mutspergene\": 0.349727797728781, \"upper_95ci_mutspergene\": 0.8576545291656879, \"legible_name\": \"RSV-A\", \"virus_family\": \"Pneumovirus\", \"color\": \"#0B194C\", \"index\": 41, \"legible_gene_name\": \"G (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"rsv_a\", \"xtick_pos\": 36, \"legible_rate_percodon\": \"1.93\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.58 muts per year\"}, {\"virus\": \"rsv\", \"subtype\": \"B\", \"virus_and_subtype\": \"rsv_B\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 2.1439093828606683, \"lower_95ci\": 1.2801227304766734, \"upper_95ci\": 3.3767220257559565, \"ci\": \"[1.2801227304766734, 3.3767220257559565]\", \"len_in_codons\": 300, \"adaptive_muts_per_year\": 0.6431728148582004, \"lower_95ci_mutspergene\": 0.384036819143002, \"upper_95ci_mutspergene\": 1.0130166077267868, \"legible_name\": \"RSV-B\", \"virus_family\": \"Pneumovirus\", \"color\": \"#6C7593\", \"index\": 41, \"legible_gene_name\": \"G (Receptor Binding)\", \"genome_type\": \"Negative-sense\", \"enveloped\": \"True\", \"url_ending\": \"rsv_b\", \"xtick_pos\": 39, \"legible_rate_percodon\": \"2.14\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.64 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_I\", \"virus_and_subtype\": \"dengue_denv1_I\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.2772247014216659, \"lower_95ci\": 0.0480079828890349, \"upper_95ci\": 0.6077131469207415, \"ci\": \"[0.04800798288903498, 0.6077131469207415]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.1372262272037246, \"lower_95ci_mutspergene\": 0.0237639515300723, \"upper_95ci_mutspergene\": 0.300818007725767, \"legible_name\": \"Dengue 1-I\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_i\", \"xtick_pos\": 42, \"legible_rate_percodon\": \"0.28\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.14 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_III\", \"virus_and_subtype\": \"dengue_denv1_III\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.0840528138476497, \"lower_95ci\": 0.0, \"upper_95ci\": 0.2912781302534159, \"ci\": \"[0.0, 0.2912781302534159]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.0416061428545866, \"lower_95ci_mutspergene\": 0.0, \"upper_95ci_mutspergene\": 0.1441826744754408, \"legible_name\": \"Dengue 1-III\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_iii\", \"xtick_pos\": 43, \"legible_rate_percodon\": \"0.08\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.04 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_IV\", \"virus_and_subtype\": \"dengue_denv1_IV\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.3275670161682112, \"lower_95ci\": 0.1260494030967501, \"upper_95ci\": 0.6447779427586959, \"ci\": \"[0.12604940309675014, 0.6447779427586959]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.1621456730032646, \"lower_95ci_mutspergene\": 0.0623944545328913, \"upper_95ci_mutspergene\": 0.3191650816655545, \"legible_name\": \"Dengue 1-IV\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_iv\", \"xtick_pos\": 44, \"legible_rate_percodon\": \"0.33\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.16 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv1_V\", \"virus_and_subtype\": \"dengue_denv1_V\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": -0.0095111020473242, \"lower_95ci\": -0.0304944848967084, \"upper_95ci\": 0.0, \"ci\": \"[-0.03049448489670842, 0.0]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": -0.0047079955134255, \"lower_95ci_mutspergene\": -0.0150947700238706, \"upper_95ci_mutspergene\": 0.0, \"legible_name\": \"Dengue 1-V\", \"virus_family\": \"Flavivirus\", \"color\": \"#0db0d6\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv1_v\", \"xtick_pos\": 45, \"legible_rate_percodon\": \"-0.01\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"-0.0 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_AA\", \"virus_and_subtype\": \"dengue_denv2_AA\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.3356339849242723, \"lower_95ci\": 0.0342444878870831, \"upper_95ci\": 0.8756110481375577, \"ci\": \"[0.03424448788708317, 0.8756110481375577]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.1661388225375148, \"lower_95ci_mutspergene\": 0.0169510215041061, \"upper_95ci_mutspergene\": 0.4334274688280911, \"legible_name\": \"Dengue 2-AA\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_aa\", \"xtick_pos\": 48, \"legible_rate_percodon\": \"0.34\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.17 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_AI\", \"virus_and_subtype\": \"dengue_denv2_AI\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.231380061824413, \"lower_95ci\": 0.0231912292530623, \"upper_95ci\": 0.466973072382025, \"ci\": \"[0.023191229253062343, 0.466973072382025]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.1145331306030844, \"lower_95ci_mutspergene\": 0.0114796584802658, \"upper_95ci_mutspergene\": 0.2311516708291023, \"legible_name\": \"Dengue 2-AI\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_ai\", \"xtick_pos\": 49, \"legible_rate_percodon\": \"0.23\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.11 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv2_C\", \"virus_and_subtype\": \"dengue_denv2_C\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": -0.0643750628828253, \"lower_95ci\": -0.1735995417057223, \"upper_95ci\": 0.081578958446469, \"ci\": \"[-0.17359954170572234, 0.08157895844646901]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": -0.0318656561269985, \"lower_95ci_mutspergene\": -0.0859317731443325, \"upper_95ci_mutspergene\": 0.0403815844310021, \"legible_name\": \"Dengue 2-C\", \"virus_family\": \"Flavivirus\", \"color\": \"#2cccf2\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv2_c\", \"xtick_pos\": 50, \"legible_rate_percodon\": \"-0.06\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"-0.03 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv3_II\", \"virus_and_subtype\": \"dengue_denv3_II\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.1910357962579551, \"lower_95ci\": 0.0827209609842614, \"upper_95ci\": 0.3886853493722831, \"ci\": \"[0.08272096098426143, 0.38868534937228316]\", \"len_in_codons\": 493, \"adaptive_muts_per_year\": 0.0941806475551718, \"lower_95ci_mutspergene\": 0.0407814337652408, \"upper_95ci_mutspergene\": 0.1916218772405356, \"legible_name\": \"Dengue 3-II\", \"virus_family\": \"Flavivirus\", \"color\": \"#76def7\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv3_ii\", \"xtick_pos\": 53, \"legible_rate_percodon\": \"0.19\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.09 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv3_III\", \"virus_and_subtype\": \"dengue_denv3_III\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.0317617470437703, \"lower_95ci\": -0.0001068335658367, \"upper_95ci\": 0.1281931768446245, \"ci\": \"[-0.0001068335658367902, 0.1281931768446245]\", \"len_in_codons\": 493, \"adaptive_muts_per_year\": 0.0156585412925788, \"lower_95ci_mutspergene\": -5.2668947957537565e-05, \"upper_95ci_mutspergene\": 0.0631992361843998, \"legible_name\": \"Dengue 3-III\", \"virus_family\": \"Flavivirus\", \"color\": \"#76def7\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv3_iii\", \"xtick_pos\": 54, \"legible_rate_percodon\": \"0.03\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.02 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv4_I\", \"virus_and_subtype\": \"dengue_denv4_I\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.4061259877770878, \"lower_95ci\": 0.1871819341848133, \"upper_95ci\": 0.6295197691163875, \"ci\": \"[0.1871819341848133, 0.6295197691163875]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.2010323639496584, \"lower_95ci_mutspergene\": 0.0926550574214825, \"upper_95ci_mutspergene\": 0.3116122857126118, \"legible_name\": \"Dengue 4-I\", \"virus_family\": \"Flavivirus\", \"color\": \"#adebfa\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv4_i\", \"xtick_pos\": 57, \"legible_rate_percodon\": \"0.41\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.2 muts per year\"}, {\"virus\": \"dengue\", \"subtype\": \"denv4_II\", \"virus_and_subtype\": \"dengue_denv4_II\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 0.1246085081342652, \"lower_95ci\": 0.0161460481464703, \"upper_95ci\": 0.3103463519891937, \"ci\": \"[0.01614604814647036, 0.31034635198919375]\", \"len_in_codons\": 495, \"adaptive_muts_per_year\": 0.0616812115264612, \"lower_95ci_mutspergene\": 0.0079922938325028, \"upper_95ci_mutspergene\": 0.1536214442346509, \"legible_name\": \"Dengue 4-II\", \"virus_family\": \"Flavivirus\", \"color\": \"#adebfa\", \"index\": 35, \"legible_gene_name\": \"E (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"True\", \"url_ending\": \"denv4_ii\", \"xtick_pos\": 58, \"legible_rate_percodon\": \"0.12\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.06 muts per year\"}, {\"virus\": \"enterovirusd68\", \"subtype\": null, \"virus_and_subtype\": \"enterovirusd68\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.517049084058603, \"lower_95ci\": 0.7582093757458963, \"upper_95ci\": 2.327216254157004, \"ci\": \"[0.7582093757458963, 2.327216254157004]\", \"len_in_codons\": 309, \"adaptive_muts_per_year\": 0.4687681669741083, \"lower_95ci_mutspergene\": 0.2342866971054819, \"upper_95ci_mutspergene\": 0.7191098225345142, \"legible_name\": \"Enterovirus D68\", \"virus_family\": \"Picornavirus\", \"color\": \"#a373a5\", \"index\": 32, \"legible_gene_name\": \"VP1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"False\", \"url_ending\": \"enterovirusd68\", \"xtick_pos\": 64, \"legible_rate_percodon\": \"1.52\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.47 muts per year\"}, {\"virus\": \"norovirus\", \"subtype\": null, \"virus_and_subtype\": \"norovirus\", \"gene\": \"receptor_binding\", \"adaptive_subs_per_codon_per_year\": 1.81000833263981, \"lower_95ci\": 1.1967182413852349, \"upper_95ci\": 2.309784031755586, \"ci\": \"[1.1967182413852349, 2.309784031755586]\", \"len_in_codons\": 541, \"adaptive_muts_per_year\": 0.9792145079581372, \"lower_95ci_mutspergene\": 0.647424568589412, \"upper_95ci_mutspergene\": 1.249593161179772, \"legible_name\": \"Norovirus GII.4\", \"virus_family\": \"Caliciviridae\", \"color\": \"#d0b9d2\", \"index\": 14, \"legible_gene_name\": \"VP1 (Receptor Binding)\", \"genome_type\": \"Positive-sense\", \"enveloped\": \"False\", \"url_ending\": \"norovirus_gii4\", \"xtick_pos\": 67, \"legible_rate_percodon\": \"1.81\\u00d710\\u207b\\u00b3 muts per codon per year\", \"legible_rate_pergene\": \"0.98 muts per year\"}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.HConcatChart(...)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#set height and width of plots\n",
    "standard_width = 800\n",
    "standard_height = 250\n",
    "\n",
    "\n",
    "# generate the points\n",
    "points_p = alt.Chart(df_polymerase).transform_calculate(\n",
    "    url='https://blab.github.io/atlas-of-viral-adaptation/' + alt.datum.url_ending\n",
    ").mark_point(\n",
    "    filled=True,\n",
    "    size=200, \n",
    "    opacity=1\n",
    ").encode(\n",
    "    x=alt.X('xtick_pos:Q', title='', \n",
    "            axis=alt.Axis(values = all_xticks, labelExpr=label_expr_str, tickCount=len(all_xticks))),\n",
    "    y=alt.Y('adaptive_muts_per_year:Q', title='', axis=alt.Axis(format=\".1f\")),\n",
    "    color=alt.Color('legible_name', scale=alt.Scale(domain=dom, range=rng), legend=None),\n",
    "    href='url:N',\n",
    "    tooltip=[alt.Tooltip('legible_name:N', title='Virus'),\n",
    "             alt.Tooltip('virus_family:N', title='Family'),\n",
    "             alt.Tooltip('genome_type:N', title='Genome Type'),\n",
    "             alt.Tooltip('enveloped:N', title='Enveloped'),\n",
    "             alt.Tooltip('legible_gene_name:N', title='Protein/Subunit'), \n",
    "             alt.Tooltip('legible_rate_pergene', title='Rate')],\n",
    ").properties(\n",
    "    width=standard_width,\n",
    "    height=standard_height*p_relative_range, \n",
    "    title='Polymerase'\n",
    ")\n",
    "\n",
    "points_r = alt.Chart(df_receptorbinding).transform_calculate(\n",
    "    url='https://blab.github.io/atlas-of-viral-adaptation/' + alt.datum.url_ending\n",
    ").mark_point(\n",
    "    filled=True,\n",
    "    size=200, opacity=1,\n",
    ").encode(\n",
    "    x=alt.X('xtick_pos:Q', title='', \n",
    "            axis=alt.Axis(values = all_xticks, labelExpr=label_expr_str, tickCount=len(all_xticks))),\n",
    "    y=alt.Y('adaptive_muts_per_year:Q', title='', axis=alt.Axis(format=\".1f\")),\n",
    "    color=alt.Color('legible_name', scale=alt.Scale(domain=dom, range=rng), legend=None),\n",
    "    href='url:N',\n",
    "    tooltip=[alt.Tooltip('legible_name:N', title='Virus'),\n",
    "             alt.Tooltip('virus_family:N', title='Family'),\n",
    "             alt.Tooltip('genome_type:N', title='Genome Type'),\n",
    "             alt.Tooltip('enveloped:N', title='Enveloped'),\n",
    "             alt.Tooltip('legible_gene_name:N', title='Protein/Subunit'), \n",
    "             alt.Tooltip('legible_rate_pergene', title='Rate')],\n",
    ").properties(\n",
    "    width=standard_width,\n",
    "    height=standard_height, \n",
    "    title='Receptor-Binding'\n",
    ")\n",
    "\n",
    "#generate the error bars\n",
    "errorbars_p = alt.Chart(df_polymerase).mark_errorbar().encode(\n",
    "    x=alt.X('xtick_pos:Q', sort=virus_order, title='', axis=alt.Axis(labels=False)),\n",
    "    y=alt.Y(\"lower_95ci_mutspergene\", title=''),\n",
    "    y2=\"upper_95ci_mutspergene\", \n",
    "    color=alt.Color('legible_name', scale=alt.Scale(domain=dom, range=rng), legend=None),\n",
    "    tooltip=alt.value(None)\n",
    ").properties(\n",
    "    width=standard_width,\n",
    "    height=standard_height*p_relative_range\n",
    ")\n",
    "\n",
    "\n",
    "errorbars_r = alt.Chart(df_receptorbinding).mark_errorbar().encode(\n",
    "    x=alt.X('xtick_pos:Q', sort=virus_order, title=''),\n",
    "    y=alt.Y(\"lower_95ci_mutspergene\", title=''),\n",
    "    y2=\"upper_95ci_mutspergene\",\n",
    "    color=alt.Color('legible_name', scale=alt.Scale(domain=dom, range=rng), legend=None),\n",
    "    tooltip=alt.value(None)\n",
    ").properties(\n",
    "    width=standard_width,\n",
    "    height=standard_height\n",
    ")\n",
    "\n",
    "#vertical line to divide enveloped from non-enveloped viruses\n",
    "vert_divider = alt.Chart(pd.DataFrame({'x': [enveloped_vert_divider]})).mark_rule(strokeDash=[3,5]).encode(x='x')\n",
    "enveloped_text = vert_divider.mark_text(\n",
    "        align='right',\n",
    "        color='#6e6e6e',\n",
    "        dy=-120,\n",
    "        dx=-10  # Nudges text to left so it doesn't appear on top of the bar. in pixels\n",
    ").encode(\n",
    "    text=alt.value(['enveloped', 'viruses']))\n",
    "nonenveloped_text = vert_divider.mark_text(\n",
    "        align='left',\n",
    "        color='#6e6e6e',\n",
    "        dy=-120,\n",
    "        dx=10  # Nudges text to left so it doesn't appear on top of the bar. in pixels\n",
    ").encode(\n",
    "    text=alt.value(['non-enveloped', 'viruses']))\n",
    "\n",
    "\n",
    "\n",
    "#y-axis label for both plots\n",
    "text = alt.Chart().mark_text(\n",
    "    align=\"center\",\n",
    "    baseline=\"bottom\",\n",
    "    fontSize=16,\n",
    "    fontWeight=400,\n",
    "    angle=270,\n",
    "    color='black'\n",
    ").encode(\n",
    "    x=alt.value(10),  # pixels from left\n",
    "    y=alt.value(140),  # pixels from top\n",
    "    text=alt.value(['Rate of Adaptation'])\n",
    ")\n",
    "\n",
    "#y-axis label for both plots\n",
    "text2 = alt.Chart().mark_text(\n",
    "    align=\"center\",\n",
    "    baseline=\"bottom\",\n",
    "    fontSize=12,\n",
    "    fontWeight=400,\n",
    "    angle=270,\n",
    "    color='black'\n",
    ").encode(\n",
    "    x=alt.value(0),  # pixels from left\n",
    "    y=alt.value(140),  # pixels from top\n",
    "    text=alt.value(['Adaptive Mutations per Year'])\n",
    ")\n",
    "\n",
    "#layer the points and error bars\n",
    "top = alt.layer(points_p, errorbars_p, vert_divider, data=df_polymerase)\n",
    "bottom = alt.layer(points_r, errorbars_r, vert_divider, enveloped_text, nonenveloped_text, data=df_receptorbinding)\n",
    "\n",
    "#stack the genes\n",
    "plot_layout = alt.vconcat(top, bottom, spacing=10)\n",
    "\n",
    "#add the yaxis label\n",
    "chart = alt.hconcat(text, text2, plot_layout, spacing=0).configure_axis(\n",
    "    grid=False\n",
    ").configure_view(\n",
    "    strokeWidth=0\n",
    ").configure_title(\n",
    "    fontSize=18,\n",
    "    anchor='start',\n",
    "    color='#c6c6c6'\n",
    ").resolve_scale(\n",
    "    y='independent',\n",
    "    x='independent',\n",
    ").configure_axisX(labelAngle=270)\n",
    "\n",
    "chart.save('../atlas-of-viral-adaptation/assets/adaptation_pergene_overview_plot.html')\n",
    "\n",
    "chart\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfa36ce5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d854f06",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "821428d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "356a018b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9722ed72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebed3301",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
