{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "89396633",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import json\n",
    "import random\n",
    "import ast\n",
    "import re\n",
    "import os\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation, CompoundLocation\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Align import AlignInfo\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.lines as mlines\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15fbc0ff",
   "metadata": {},
   "source": [
    "## Count fixations at each site, using same time windows as adaptation analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1aed8993",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_virus_config(virus):\n",
    "    config_json = f'config/adaptive_evo_config_{virus}.json'\n",
    "    with open(config_json) as json_handle:\n",
    "        configs = json.load(json_handle)\n",
    "        \n",
    "    return configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4c1ea580",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_data(saved_data_name):\n",
    "    with open(saved_data_name) as json_handle:\n",
    "        json_dict = json.load(json_handle)\n",
    "        (aligned_codons_in_window, outgroup_codons, year_windows) = (json_dict['aligned_codons_in_window'], \n",
    "                                                                     json_dict['outgroup_codons'], \n",
    "                                                                     json_dict['year_windows'])\n",
    "    return aligned_codons_in_window, outgroup_codons, year_windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc224036",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_alignment_outgroup_data(virus, subtype, gene, window, min_seqs, year_max, year_min):\n",
    "    \"\"\"\n",
    "    If the alignment and outgroup data has already been organized for this \n",
    "    virus, subtype, gene, window, min_seqs, year_min, year_max combo, then read in the data. \n",
    "    Otherwise, calculate it and save it. Return aligned_codons_in_window, outgroup_codons, year_windows\n",
    "    \"\"\"\n",
    "    \n",
    "    filepath = 'adaptation_results/intermediates/'\n",
    "    \n",
    "    saved_data_name = f'{filepath}input_data_{virus}_{subtype}_{gene}_{window}_{min_seqs}_{year_min}_{year_max}.json'\n",
    "    \n",
    "    #prepare data and save it if this has not already been done\n",
    "    if not path.exists(saved_data_name):\n",
    "        print(f'need to run adaptation analysis for {virus} {gene}')\n",
    "    \n",
    "    #read in the prepared alignment and outgroup data\n",
    "    (aligned_codons_in_window, outgroup_codons, year_windows) = readin_data(saved_data_name)\n",
    "    \n",
    "    #reformat alignment outgroup to flat list of codons\n",
    "    outgroup_codons = [x[0] for x in outgroup_codons]\n",
    "\n",
    "    return aligned_codons_in_window, outgroup_codons, year_windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "080363c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def walk_through_sites(outgroup_codons, alignment_seqs, midfreq_high):\n",
    "    \"\"\"\n",
    "    Walk through each codon in each sequence of the alignment and determine whether \n",
    "    there was a synonymous mut, nonsynonymous mut or neither. \n",
    "    If there was a mutation, want to know its nucleotide position\n",
    "    \"\"\"\n",
    "\n",
    "    \n",
    "    #initialize arrays to count fixations at each codon \n",
    "    nonsyn_fixations_in_window = np.zeros(len(outgroup_codons))\n",
    "    syn_fixations_in_window = np.zeros(len(outgroup_codons))\n",
    "    \n",
    "    #initialize arrays to count fixations at each nucleotide\n",
    "    nonsyn_fixations_in_window_nt = np.zeros(len(outgroup_codons)*3)\n",
    "    syn_fixations_in_window_nt = np.zeros(len(outgroup_codons)*3)\n",
    "    \n",
    "    #initialize dictionary to keep track of codons {pos:codon seq} that have fixed during this time window, \n",
    "    #so that outgroup_seq can be updated\n",
    "    fixed_codons = {}\n",
    "    \n",
    "    #list of codons in alignment organized as [['ATG','ATG','ATG'],['TAG','TAG','TAG']]\n",
    "    alignment_codons = [[] for x in outgroup_codons]\n",
    "    for isolate_codons in alignment_seqs:\n",
    "        for i in range(len(isolate_codons)):\n",
    "            alignment_codons[i].append(str(isolate_codons[i]))  \n",
    "\n",
    "    \n",
    "    #walk through sequence codon by codon\n",
    "    for i in range(len(outgroup_codons)):\n",
    "        #only consider unambiguous sequencing\n",
    "        if set('AGCT').intersection(set(str(outgroup_codons[i]))) == set(str(outgroup_codons[i])):\n",
    "            #find fixations or near-fixations\n",
    "            fixation_type, fixed_codon_seq, fixed_nts = find_mutation_fixations(i, outgroup_codons[i], \n",
    "                                                                     alignment_codons[i], midfreq_high)\n",
    "            if fixed_codon_seq!=None:\n",
    "                fixed_codons[i] = fixed_codon_seq\n",
    "            if fixation_type== 'nonsynonymous':\n",
    "                nonsyn_fixations_in_window[i]+=1\n",
    "                nonsyn_fixations_in_window_nt[i:i+3] =fixed_nts\n",
    "            elif fixation_type== 'synonymous':\n",
    "                syn_fixations_in_window[i]+=1\n",
    "                syn_fixations_in_window_nt[i:i+3] = fixed_nts\n",
    "                \n",
    "                    \n",
    "    return (nonsyn_fixations_in_window, syn_fixations_in_window, fixed_codons, \n",
    "            nonsyn_fixations_in_window_nt, syn_fixations_in_window_nt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8c360792",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_mutation_fixations(pos, outgroup_codon, alignment_codons, midfreq_high):\n",
    "    \"\"\"\n",
    "    At a given codon position, find whether any mutations occurred, \n",
    "    and if so what frequency they are at in the population\n",
    "    \n",
    "    Classify mutations that reach between midfreq_high and 100% as fixations\n",
    "    \"\"\" \n",
    "\n",
    "    #only consider unabiguous sequencing\n",
    "    alignment_codons = [x for x in alignment_codons if set('AGCT').intersection(set(str(x))) == set(str(x))]\n",
    "\n",
    "    #get a count of the different codon sequences observed at this position\n",
    "    codon_seq_counts = Counter(alignment_codons)\n",
    "    codon_seqs = list(codon_seq_counts.keys())\n",
    "    codon_freqs = {c: (codon_seq_counts[c] / len(alignment_codons)) for c in codon_seq_counts}\n",
    "\n",
    "    \n",
    "    #all codons are the same in the alignment\n",
    "    if len(codon_seqs)==1:\n",
    "        #check if they are the same as the outgroup\n",
    "        #or whether they are a fixed mutation\n",
    "        if outgroup_codon==codon_seqs[0]:\n",
    "            site_type = 'no_fixation'\n",
    "        #or whether they are a fixed mutation\n",
    "        elif outgroup_codon!=codon_seqs[0]:\n",
    "            site_type = 'fixation'\n",
    "            fixed_codon = codon_seqs[0]\n",
    "            \n",
    "    #if there are multiple codon sequences observed at this position, \n",
    "    #see if there are any mutations present at a frequency of midfreq_high or higher\n",
    "    elif len(codon_seqs)!=1:\n",
    "        #default is no fixation, can be overwritten if one codon is present at high freq and is a mutation\n",
    "        site_type = 'no_fixation'\n",
    "        for cod in codon_seqs:\n",
    "            #check whether any of the codon sequences are present at high enough frequency\n",
    "            if codon_freqs[cod] >= midfreq_high:\n",
    "                #check if it is a mutation\n",
    "                if cod != outgroup_codon:\n",
    "                    site_type = 'fixation'\n",
    "                    fixed_codon = cod\n",
    "\n",
    "\n",
    "    # check if fixation was synonymous or nonsynonymous                \n",
    "    if site_type == 'fixation':\n",
    "        outgroup_aa = Seq(outgroup_codon).translate()\n",
    "        alignment_codon_aa = Seq(fixed_codon).translate()\n",
    "        if outgroup_aa == alignment_codon_aa:\n",
    "            fixation_type = 'synonymous'\n",
    "        elif outgroup_aa != alignment_codon_aa:\n",
    "            fixation_type = 'nonsynonymous'\n",
    "        #find which nt positions have fixed\n",
    "        fixed_nts = [1 if outgroup_codon[i]!=fixed_codon[i] else 0 for i in range(len(outgroup_codon))]\n",
    "    elif site_type == 'no_fixation':\n",
    "        fixation_type = None\n",
    "        fixed_codon = None\n",
    "        fixed_nts = [0,0,0]\n",
    "    \n",
    "    return fixation_type, fixed_codon, fixed_nts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "34fdbde1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_codon_fixations(alignment_sequences, outgroup_codons, midfreq_high):\n",
    "    \"\"\"\n",
    "    For each time window, find if any mutations have fixed \n",
    "    (or nearly fixed, meaning reach a frequency exceeding midfreq_high).\n",
    "    Find whether those mutations are nonsynonymous or synonymous.\n",
    "    Keep a count of fixations/near-fixations at each codon\n",
    "    \"\"\"\n",
    "    \n",
    "\n",
    "\n",
    "    #initiate arrays to record fixations (or near-fixations) at all codons in alignment\n",
    "    #keep track of this for both nonsynonymous and synonymous mutations\n",
    "    nonsynonymous_fixations = np.zeros(len(outgroup_codons))\n",
    "    synonymous_fixations = np.zeros(len(outgroup_codons))\n",
    "    \n",
    "    nonsynonymous_fixations_nt = np.zeros(len(outgroup_codons)*3)\n",
    "    synonymous_fixations_nt = np.zeros(len(outgroup_codons)*3)\n",
    "    \n",
    "    nonsyn_nt_fixations_by_year = {}\n",
    "        \n",
    "\n",
    "    for years, alignment_seqs in alignment_sequences.items():\n",
    "\n",
    "  \n",
    "        #look for fixations in each time window\n",
    "        (nonsyn_fixations_in_window, syn_fixations_in_window, fixed_codons, \n",
    "            nonsyn_fixations_in_window_nt, syn_fixations_in_window_nt) = walk_through_sites(outgroup_codons, \n",
    "                                                                                                alignment_seqs, midfreq_high)\n",
    "        nonsynonymous_fixations += nonsyn_fixations_in_window\n",
    "        synonymous_fixations += syn_fixations_in_window\n",
    "        \n",
    "        nonsynonymous_fixations_nt +=nonsyn_fixations_in_window_nt\n",
    "        synonymous_fixations_nt +=syn_fixations_in_window_nt\n",
    "        \n",
    "        #keep track of how many nt fixations there are in each year window (by window midpoint)\n",
    "        window_midpoint = (int(years.split('-')[0])+int(years.split('-')[1]))/2\n",
    "        nonsyn_nt_fixations_by_year[window_midpoint] = sum(nonsyn_fixations_in_window_nt)\n",
    "        \n",
    "        #update outgroup_seq at codons that have fixed mutations \n",
    "        for pos, seq in fixed_codons.items():\n",
    "            outgroup_codons[pos] = seq\n",
    "    \n",
    "\n",
    "    return (nonsynonymous_fixations, synonymous_fixations, \n",
    "            nonsynonymous_fixations_nt, nonsyn_nt_fixations_by_year, nonsyn_nt_fixations_by_year)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "405b7b59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_fixations(virus, subtype, gene, \n",
    "         window=5, min_seqs=3, year_max=None, year_min=None, midfreq_high=1.0):\n",
    "    \"\"\"\n",
    "    Count the number of fixations at each nucleotide site\n",
    "    \n",
    "    Return the total length of the gene, the total number of sites with at least one nonsyn mutation, \n",
    "    and the total number of sites with multiple nonsyn mutations\n",
    "    \"\"\"\n",
    "    #for viruses with under 12 years of data, the adaptation analysis uses smaller windows\n",
    "    if virus in ['h1n1pdm', 'mumps']:\n",
    "        window, min_seqs = 3,3\n",
    "    \n",
    "    (aligned_codons_in_window, outgroup_codons, year_windows) = get_alignment_outgroup_data(virus, subtype, gene, \n",
    "                                                                                            window, min_seqs, \n",
    "                                                                                            year_max, year_min)\n",
    "    \n",
    "    total_len_gene = len(outgroup_codons)*3\n",
    "    \n",
    "    (nonsynonymous_fixations, synonymous_fixations,\n",
    "     nonsynonymous_fixations_nt, synonymous_fixations_nt, \n",
    "     nonsyn_nt_fixations_by_year) = count_codon_fixations(aligned_codons_in_window, \n",
    "                                                          outgroup_codons, midfreq_high)\n",
    "    \n",
    "\n",
    "    #tally the number of sites with fixations (any number)\n",
    "    codons_w_nonsyn_fixations = np.count_nonzero(nonsynonymous_fixations)\n",
    "    codons_w_syn_fixations = np.count_nonzero(synonymous_fixations)\n",
    "    nt_sites_w_a_nonsyn_fixation = np.count_nonzero(nonsynonymous_fixations_nt)\n",
    "    \n",
    "    #tally the number of sites with more than one nonsynonymous fixations\n",
    "    codons_w_multiple_nonsyn_fixations = len([x for x in nonsynonymous_fixations if x>=2])\n",
    "    codons_w_multiple_syn_fixations = len([x for x in synonymous_fixations if x>=2])\n",
    "    nts_w_multiple_nonsyn_fixations = len([x for x in nonsynonymous_fixations_nt if x>=2])\n",
    "    nts_w_multiple_syn_fixations = len([x for x in synonymous_fixations_nt if x>=2])\n",
    "    \n",
    "    print(f'gene length: {total_len_gene}')\n",
    "    print(f'Nt sites with at least one nonsyn fixation: {nt_sites_w_a_nonsyn_fixation}')\n",
    "    print(f'Nt sites with only one nonsyn fixation: {nt_sites_w_a_nonsyn_fixation - nts_w_multiple_nonsyn_fixations}')\n",
    "    print(f'Nt sites with multiple nonsyn fixation: {nts_w_multiple_nonsyn_fixations}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "afc41d56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene length: 1635\n",
      "Nt sites with at least one nonsyn fixation: 9\n",
      "Nt sites with only one nonsyn fixation: 9\n",
      "Nt sites with multiple nonsyn fixation: 0\n"
     ]
    }
   ],
   "source": [
    "count_fixations('229e', None, 'S1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "948eaf4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene length: 987\n",
      "Nt sites with at least one nonsyn fixation: 52\n",
      "Nt sites with only one nonsyn fixation: 44\n",
      "Nt sites with multiple nonsyn fixation: 8\n"
     ]
    }
   ],
   "source": [
    "count_fixations('h3n2', None, 'ha1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b584ab14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene length: 2259\n",
      "Nt sites with at least one nonsyn fixation: 11\n",
      "Nt sites with only one nonsyn fixation: 9\n",
      "Nt sites with multiple nonsyn fixation: 2\n"
     ]
    }
   ],
   "source": [
    "count_fixations('oc43', 'a', 'S1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1e6af1f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
