{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2db21703",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import math\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.lines import Line2D  \n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from augur.utils import json_to_tree\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d9b77b6",
   "metadata": {},
   "source": [
    "## Figure 5: mutations in the receptor-binding region of SARS-CoV-2 are accumulating much faster than in other endemic viruses\n",
    "\n",
    "Plot the accumulation of mutations in the receptor-binding protein for SARS2 over time, for Omicron-only and for all 10 other viruses that evolve antigenically "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "339142ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_virus_config(virus):\n",
    "    config_json = f'config/adaptive_evo_config_{virus}.json'\n",
    "\n",
    "    with open(config_json) as json_handle:\n",
    "        configs = json.load(json_handle)\n",
    "        \n",
    "    return configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed0570c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in colors from the manuscript_colors config file\n",
    "color_config_json = f'config/manuscript_colors.json'\n",
    "with open(color_config_json) as json_handle:\n",
    "    colors = json.load(json_handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7cf19038",
   "metadata": {},
   "outputs": [],
   "source": [
    "#path to the tree json files\n",
    "tree_paths = {'sars2': '../sars2/auspice/ncov_gisaid_global_all-time.json', \n",
    "              'sars2_21L': '../sars2_21L/auspice/ncov_gisaid_21L_global_all-time.json', \n",
    "              'h3n2': \"../seasonal-flu/auspice/flu_seasonal_h3n2_ha_60y.json\", \n",
    "              'h1n1pdm': \"../seasonal-flu/auspice/flu_seasonal_h1n1pdm_ha_12y.json\", \n",
    "              'vic': \"../seasonal-flu/new_builds/auspice/flu_seasonal_vic_ha_60y.json\", \n",
    "              'yam': \"../seasonal-flu/auspice/flu_seasonal_yam_ha_60y.json\", \n",
    "              '229e': \"../seasonal-covs/229e/auspice/229e_S1.json\", \n",
    "              'oc43_A': \"../seasonal-covs/oc43/separate_lineages/auspice/seasonal_corona_oc43_a_s1.json\", \n",
    "              'norovirus': \"../norovirus/auspice/norovirus_GII.4_VP1.json\", \n",
    "              'enterovirusd68': \"../enterovirus_d68/genome/enterovirus_d68_genome.json\", \n",
    "              'rsv_A': \"../rsv/rsv_step3/auspice/rsv_A.json\", \n",
    "              'rsv_B': \"../rsv/rsv_step3/auspice/rsv_B.json\"}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7c741cc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#name of RB protein\n",
    "rb_protein = {'sars2':'S1', 'sars2_21L':'S1', \n",
    "              'h3n2': 'HA1', 'h1n1pdm': 'HA1', \n",
    "              'vic':'HA1', 'yam': 'HA1', \n",
    "              '229e': 'S1', 'oc43_A': 's1', \n",
    "              'norovirus': \"VP1\", 'enterovirusd68': \"VP1\", \n",
    "              'rsv_A': \"G\", 'rsv_B': \"G\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e601e952",
   "metadata": {},
   "outputs": [],
   "source": [
    "#length of RB protein\n",
    "rb_length = {'sars2':671, 'sars2_21L':671,\n",
    "             'h3n2': 328, 'h1n1pdm': 326, 'vic':346, 'yam': 346, \n",
    "             '229e': 544, 'oc43_A': 752, \n",
    "             'norovirus': 540, 'enterovirusd68': 308, \n",
    "             'rsv_A': 321, 'rsv_B': 319}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d6a86c93",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_mutation_accumulation_on_tips(virus):\n",
    "    \"\"\"\n",
    "    For each tip in the tree, find the nonsynonymous mutations in the receptor-binding protein \n",
    "    that have accumulated between the root of the tree and the tip\n",
    "    \"\"\"\n",
    "    \n",
    "    tree_file = tree_paths[virus]\n",
    "    rb = rb_protein[virus]\n",
    "    \n",
    "    #read in the tree\n",
    "    with open(tree_file, 'r') as f:\n",
    "        tree_json = json.load(f)\n",
    "\n",
    "    #put tree in Bio.phylo format\n",
    "    tree = json_to_tree(tree_json)\n",
    "    \n",
    "    #initiate dictionary to keep track of the date and number of accumulated mutations for each tip on the tree\n",
    "    aa_muts_over_time = []\n",
    "    \n",
    "    #get date of the root\n",
    "    root_date = tree.root.node_attrs['num_date']['value']\n",
    "    \n",
    "    for node in tree.find_clades(terminal=True):\n",
    "        path = tree.get_path(node)\n",
    "#         print(node.branch_attrs['mutations'])\n",
    "#         14-685\n",
    "        \n",
    "        #count the number of amino acid mutations between root and tip\n",
    "        #sars2 S1 is not labeled, instead will get S1 mutations from spike mutations (S1 is spike AAs 14-685)\n",
    "        if virus in ['sars2', 'sars2_21L']:\n",
    "            muts_in_spike = [x.branch_attrs['mutations']['S'] for x in path if 'S' in x.branch_attrs['mutations'].keys()]            \n",
    "            muts_in_s1 = [[y for y in x if 14<=int(y[1:-1])<=685] for x in muts_in_spike]\n",
    "            muts_in_s1_no_deletions = [[y for y in x if '-' not in y] for x in muts_in_s1]\n",
    "            count_muts_in_s1 = sum([len(x) for x in muts_in_s1])\n",
    "            count_muts_in_s1_no_deletions = sum([len(x) for x in muts_in_s1_no_deletions])\n",
    "            muts_on_path = count_muts_in_s1_no_deletions\n",
    "        else:\n",
    "            muts_in_rb = sum([len(x.branch_attrs['mutations'][rb]) for x in path if rb in x.branch_attrs['mutations'].keys()])\n",
    "            muts_in_rb_no_deletions = sum([len([y for y in x.branch_attrs['mutations'][rb] if '-' not in y]) for x in path if rb in x.branch_attrs['mutations'].keys()])\n",
    "            muts_on_path = muts_in_rb_no_deletions\n",
    "            \n",
    "        muts_on_path_per_codon = muts_on_path/rb_length[virus]\n",
    "        #get the date of this tip\n",
    "        num_date = node.node_attrs['num_date']['value']\n",
    "        #time between root and tip\n",
    "        time = num_date - root_date\n",
    "        \n",
    "        aa_muts_over_time.append({'virus':virus, 'tip': node.name, \n",
    "                                  'date': num_date, 'time_from_root': time, \n",
    "                                  'aa_muts': muts_on_path, 'aa_muts_per_codon': muts_on_path_per_codon})\n",
    "    \n",
    "    df = pd.DataFrame(aa_muts_over_time)\n",
    "    \n",
    "    return df    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "150a8d28",
   "metadata": {},
   "outputs": [],
   "source": [
    "#legible virus names for plotting\n",
    "virus_name_legible = {'sars2': 'SARS-CoV-2', 'sars2_21L': 'SARS-CoV-2, Omicron 21L',\n",
    "                      'h3n2': 'A/H3N2', 'h1n1pdm': 'A/H1N1pdm', 'vic':'B/Vic', 'yam': 'B/Yam', \n",
    "                      '229e': 'CoV 229E', 'oc43_A': 'CoV OC43-A', \n",
    "                      'norovirus': 'Norovirus GII.4', 'enterovirusd68': 'Enterovirus D68', \n",
    "                      'rsv_A': 'RSV-A', 'rsv_B': 'RSV-B'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3b7178bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save virus data so plots will run faster\n",
    "all_viruses = ['sars2', 'sars2_21L', 'h3n2', 'h1n1pdm', 'vic', 'yam', \n",
    "               '229e', 'oc43_A', 'enterovirusd68', 'norovirus',\n",
    "               'rsv_A', 'rsv_B']\n",
    "\n",
    "virus_data = {}\n",
    "for virus in all_viruses:\n",
    "    df = count_mutation_accumulation_on_tips(virus)\n",
    "    virus_data[virus] = df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fec7a9c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mut_accumulation_vs_time(viruses, filename=None):\n",
    "    \"\"\"\n",
    "    For each virus, plot every tip as the number of accumulated mutations vs. time since root\n",
    "    \"\"\"\n",
    "    \n",
    "    #plot styling\n",
    "    sns.set_style(\"white\")\n",
    "    plt.rcParams['axes.labelsize'] = 10\n",
    "    plt.rcParams['axes.labelweight'] = 'normal'\n",
    "    plt.rcParams['axes.titlesize'] = 12\n",
    "    plt.rcParams['xtick.labelsize'] = 12\n",
    "    plt.rcParams['ytick.labelsize'] = 12\n",
    "    plt.rcParams['figure.dpi']= 300\n",
    "    \n",
    "    fig, ((ax1,ax2,ax3), (ax4,ax5,ax6), \n",
    "          (ax7,ax8,ax9), (ax10,ax11,ax12)) = plt.subplots(4,3, sharex=False, sharey=False, figsize=(18,12), \n",
    "                                                          gridspec_kw={'height_ratios': [2, 1, 1, 1]})\n",
    "    \n",
    "    fig_labels = ['A','B','C','D','E','F','G','H','I','J','K','L']\n",
    "    \n",
    "    for i in range(len(viruses)):\n",
    "        virus = viruses[i]\n",
    "        ax_i = [ax1,ax2,ax3,ax4,ax5,ax6,ax7,ax8,ax9,ax10,ax11,ax12][i]\n",
    "        \n",
    "        df = virus_data[virus]\n",
    "        #markersize depending on number of points to plot, to make things more visible\n",
    "        ms=18\n",
    "        if len(df)>2000:\n",
    "            ms = 12\n",
    "        #plot the mutation accumulation over time\n",
    "        ax_i.scatter(df['date'], df['aa_muts_per_codon'], color=colors['colors'][virus], \n",
    "                     marker='o', alpha=0.4, s=ms, zorder=1)\n",
    "\n",
    "        #get linear fit\n",
    "        coef = np.polyfit(df['date'], df['aa_muts_per_codon'], 1)\n",
    "        poly1d_fn = np.poly1d(coef)\n",
    "\n",
    "        #plot the linear regression line\n",
    "        ax_i.plot(df['date'], poly1d_fn(df['date']), '-', linewidth=2, color='#474747', zorder=2)\n",
    "        \n",
    "        slope = coef[0]\n",
    "        slope_readable = f'{round(slope*10**3, 1)}'+'$ × 10^{-3}$ subs/res/year'\n",
    "        \n",
    "        #label the virus, the rb-gene name, and the rate\n",
    "        if i in [0,1,2]:\n",
    "            ax_i.text(0.02, 0.98, virus_name_legible[virus], fontsize=20, horizontalalignment='left', \n",
    "                   verticalalignment='center', transform=ax_i.transAxes)\n",
    "            ax_i.text(0.02, 0.88, rb_protein[virus], fontsize=16, horizontalalignment='left', \n",
    "                   verticalalignment='center', transform=ax_i.transAxes)\n",
    "\n",
    "            ax_i.text(0.02, 0.8, slope_readable, fontsize=14, horizontalalignment='left', \n",
    "                   verticalalignment='center', transform=ax_i.transAxes)\n",
    "        else:\n",
    "            ax_i.text(0.02, 0.96, virus_name_legible[virus], fontsize=20, horizontalalignment='left', \n",
    "                   verticalalignment='center', transform=ax_i.transAxes)\n",
    "            ax_i.text(0.02, 0.78, rb_protein[virus].upper(), fontsize=16, horizontalalignment='left', \n",
    "                   verticalalignment='center', transform=ax_i.transAxes)\n",
    "\n",
    "            ax_i.text(0.02, 0.66, slope_readable, fontsize=14, horizontalalignment='left', \n",
    "                   verticalalignment='center', transform=ax_i.transAxes)\n",
    "        \n",
    "        #label only outer y-axes:\n",
    "        if i==0:\n",
    "            ax_i.set_ylabel('AA subs per residue\\n$(× 10^{-3})$', fontsize=14)\n",
    "        else:\n",
    "            ax_i.set_ylabel('')\n",
    "        #label only bottom x-axes\n",
    "        if i >=9:\n",
    "            ax_i.set_xlabel('Date', fontsize=14)\n",
    "        else:\n",
    "            ax_i.set_xlabel('')\n",
    "            \n",
    "        #add figure panel labels\n",
    "        ax_i.text(-0.15, 1.0, fig_labels[i], horizontalalignment='center', \n",
    "                       verticalalignment='center', weight='bold', fontsize=24, transform=ax_i.transAxes)\n",
    "\n",
    "        \n",
    "        #dynamically get the y-range by dividing x range (in years) by 45\n",
    "        x_range = df['date'].max()-df['date'].min()\n",
    "        if i in [0,1,2]:\n",
    "            ax_i.set_ylim([-0.1*(30/1000), x_range*(30/1000)])\n",
    "        #these plots are half as tall\n",
    "        else:\n",
    "            ax_i.set_ylim([-0.1*(30/1000), x_range*(30/1000)*0.5])\n",
    "        \n",
    "        #make the sars2 x-axes less crowded\n",
    "        if virus=='sars2':\n",
    "            ax_i.set_xticks([2020, 2021, 2022, 2023, 2024])\n",
    "            ax_i.set_xticklabels([2020, 2021, 2022, 2023, 2024])\n",
    "        if virus=='sars2_21L':\n",
    "            ax_i.set_xticks([2022, 2023, 2024])\n",
    "            ax_i.set_xticklabels([2022, 2023, 2024])\n",
    "            \n",
    "    plt.subplots_adjust(hspace=0.3)\n",
    "            \n",
    "        \n",
    "    sns.despine()\n",
    "    \n",
    "    if filename:\n",
    "        fig.savefig(filename, dpi=300, bbox_inches='tight', transparent=False, facecolor='white')\n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4843246a",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_mut_accumulation_vs_time(['sars2', 'sars2_21L', \n",
    "                               'h3n2', 'h1n1pdm', 'vic', 'yam', \n",
    "                               '229e', 'oc43_A', 'rsv_A', 'rsv_B', \n",
    "                               'enterovirusd68', 'norovirus'], filename = 'manuscript_figures/Figure5_SubRates.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c8af5aab",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_mut_accumulation_vs_time(['sars2', 'sars2_21L', \n",
    "                               'h3n2', 'h1n1pdm', 'vic', 'yam', \n",
    "                               '229e', 'oc43_A', 'rsv_A', 'rsv_B', \n",
    "                               'enterovirusd68', 'norovirus'], filename = 'manuscript_figures/Figure5_SubRates.tiff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78c95398",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "796a86c6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b512993",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
