{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2c44d5d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import json\n",
    "import random\n",
    "import ast\n",
    "import re\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation, CompoundLocation\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.lines as mlines\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5e7a1b6",
   "metadata": {},
   "source": [
    "### Save virus rates of adaptation and metadata info in csv for d3 to read in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "56c38b87",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_virus_config(virus):\n",
    "    \"\"\"\n",
    "    Read in the config file for this virus to get the paths to alignment, metadata files, etc \n",
    "    as well as metadata about the virus such as how many subtypes, \n",
    "    and which genes are receptor-binding or polymerase\n",
    "    \"\"\"\n",
    "    config_json = f'config/adaptive_evo_config_{virus}.json'\n",
    "    with open(config_json) as json_handle:\n",
    "        configs = json.load(json_handle)\n",
    "        \n",
    "    return configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e8ba02c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rate_in_muts_per_year(virus, subtype, gene, window, min_seqs):\n",
    "    \"\"\"\n",
    "    Add a column giving the rate of adaptation (and 95ci, if bootstrapped) in \n",
    "    adaptive muts per year (rather than per codon per year)\n",
    "    \"\"\"\n",
    "    \n",
    "    intermediates_file = f'adaptation_results/intermediates/input_data_{virus}_{subtype}_{gene}_{window}_{min_seqs}_None_None.json'\n",
    "    \n",
    "    with open(intermediates_file) as json_handle:\n",
    "        json_dict = json.load(json_handle)\n",
    "        (aligned_codons_in_window, outgroup_codons, year_windows) = (json_dict['aligned_codons_in_window'], \n",
    "                                                                     json_dict['outgroup_codons'], \n",
    "                                                                     json_dict['year_windows'])\n",
    "    \n",
    "    gene_len_in_codons = len(outgroup_codons)\n",
    "\n",
    "\n",
    "    \n",
    "    return gene_len_in_codons\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "64838590",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_rates_updated_outgroup(virus_and_subtype, gene, window, min_seqs):\n",
    "    \"\"\"\n",
    "    Read in the rates of adaptation, calculated with the update outgroup method\n",
    "    \"\"\"\n",
    "    \n",
    "    json_name = f'adaptation_results/results/{virus_and_subtype}_{gene}_{window}_{min_seqs}_adaptation_bootstrapped.json'\n",
    "\n",
    "    if path.exists(json_name):\n",
    "        with open(json_name) as json_handle:\n",
    "            json_dict = json.load(json_handle)\n",
    "            (window_midpoint, adaptive_substitutions, \n",
    "             adaptive_substitutions_per_codon, \n",
    "             rate_of_adaptation, bootstrap_adaptive_substitutions, \n",
    "             bootstrap_adaptive_substitutions_per_codon, \n",
    "             bootstrap_rate_of_adaptation) = (json_dict['window_midpoint'], \n",
    "                                              json_dict['adaptive_substitutions'], \n",
    "                                              json_dict['adaptive_substitutions_per_codon'], \n",
    "                                              json_dict['rate_of_adaptation'], \n",
    "                                              json_dict['bootstrap_adaptive_substitutions'], \n",
    "                                              json_dict['bootstrap_adaptive_substitutions_per_codon'], \n",
    "                                              json_dict['bootstrap_rate_of_adaptation'])\n",
    "            \n",
    "        slope_sci = rate_of_adaptation * (10**3)\n",
    "        bs_slope_sci = [x * (10**3) for x in bootstrap_rate_of_adaptation]\n",
    "        lower_95ci = np.percentile(sorted(bs_slope_sci), 2.5)\n",
    "        upper_95ci = np.percentile(sorted(bs_slope_sci), 97.5)\n",
    "    \n",
    "    else:\n",
    "        print(f'{json_name} does not exist')\n",
    "        \n",
    "    return slope_sci, lower_95ci, upper_95ci\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c0d8aa24",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_rates_constant_outgroup(virus_and_subtype, gene, window, min_seqs):\n",
    "    \"\"\"\n",
    "    Read in the rates of adaptation, calculated with the fixed outgroup method\n",
    "    \"\"\"\n",
    "    \n",
    "    json_name = f'bhatt_results/bhatt_results_nextstrain_{window}_{min_seqs}/{virus_and_subtype}_{gene}_bhatt_analysis_bootstrapped.json'\n",
    "\n",
    "    if path.exists(json_name):\n",
    "        with open(json_name) as json_handle:\n",
    "            json_dict = json.load(json_handle)\n",
    "            (window_midpoint, adaptive_substitutions, \n",
    "             adaptive_substitutions_per_codon, \n",
    "             rate_of_adaptation, bootstrap_adaptive_substitutions, \n",
    "             bootstrap_adaptive_substitutions_per_codon, \n",
    "             bootstrap_rate_of_adaptation) = (json_dict['window_midpoint'], \n",
    "                                              json_dict['adaptive_substitutions'], \n",
    "                                              json_dict['adaptive_substitutions_per_codon'], \n",
    "                                              json_dict['rate_of_adaptation'], \n",
    "                                              json_dict['bootstrap_adaptive_substitutions'], \n",
    "                                              json_dict['bootstrap_adaptive_substitutions_per_codon'], \n",
    "                                              json_dict['bootstrap_rate_of_adaptation'])\n",
    "            \n",
    "        slope_sci = rate_of_adaptation * (10**3)\n",
    "        bs_slope_sci = [x * (10**3) for x in bootstrap_rate_of_adaptation]\n",
    "        lower_95ci = np.percentile(sorted(bs_slope_sci), 2.5)\n",
    "        upper_95ci = np.percentile(sorted(bs_slope_sci), 97.5)\n",
    "    \n",
    "    else:\n",
    "        print(f'{json_name} does not exist')\n",
    "        \n",
    "    return slope_sci, lower_95ci, upper_95ci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6494796f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#mappers to make gene names more readable\n",
    "rota_segment_to_gene = {'A':'NSP1', 'C':'VP2', 'E':'NSP4', 'G':'VP7', 'I':'VP6', \n",
    "                        'M': 'VP3', 'N': 'NSP2', 'P':'VP4', 'R':'VP1', 'T':'NSP3'}\n",
    "\n",
    "legible_gene_names = {'3clpro':'3CLpro', 'Ha1':'HA1', 'Ha2':'HA2', \n",
    "                      'He':'HE', 'Hn':'HN', 'Na':'NA', 'Nc':'NC', \n",
    "                      'Np':'NP', 'Ns1':'NS1', 'Ns2':'NS2', 'Pa':'PA', \n",
    "                      'Pb1':'PB1', 'Pb2':'PB2', 'Rdrp':'RdRp', \n",
    "                      'Sh':'SH', 'Vp1':'VP1', 'Vp2':'VP2', 'Vp3': 'VP3', \n",
    "                      'Vp4': 'VP4', '2a':\"2A\", '2b':\"2B\", '2c':\"2C\", \n",
    "                      '3a':\"3A\", '3b':\"3B\", '3c':\"3C\", '3d':\"3D\", \n",
    "                      \"Hef1\":\"HEF1\", 'Hef2':\"HEF2\", 'Ns':\"NS\", 'E1a':'E1A', \n",
    "                      'E1b 55k':'E1B 55K', 'Iva2':'IVa2', 'Ptp':'pTP', 'Piiia':'pIIIa', '100k':'100K'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "012eaacc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gather_all_data_update_outgroup(viruses, to_include, window=5, min_seqs=3):\n",
    "    \"\"\"\n",
    "    Gather all the data for each virus and gene (with rates calculated using the update outgroup method) \n",
    "    that will be needed for the website\n",
    "    \"\"\"\n",
    "    all_data = []\n",
    "    \n",
    "    #for each virus, find whether there are any subtypes, then get all rates and metadata for these subtypes\n",
    "    for virus in viruses:\n",
    "        configs = readin_virus_config(virus)\n",
    "            \n",
    "        #metadata for virus\n",
    "        genes = configs[\"genes\"]\n",
    "        genome_type = configs['genome_type']\n",
    "        rna_dna = [x for x in ['RNA', 'DNA'] if x in configs['genome_type']][0]\n",
    "        enveloped = configs['enveloped']\n",
    "        transmission = configs['transmission']\n",
    "        cellular_receptor = configs['cellular_receptor']\n",
    "        fusion_protein_class = configs['fusion_protein_class']\n",
    "        virus_family = configs['virus_family']\n",
    "        #find the names of the polymerase and receptor-binding genes\n",
    "        polymerase = configs['polymerase']['virus_gene']\n",
    "        receptor_binding = configs['receptor_binding']['virus_gene']\n",
    "        standard_genes = {polymerase.upper():'polymerase', receptor_binding.upper():'receptor_binding'}\n",
    "        \n",
    "        #if virus has subtypes, want to get info for each subtype\n",
    "        if configs['subtype']=='True':\n",
    "            for subtype in configs['subtypes']:\n",
    "                virus_and_subtype = f'{virus}_{subtype}'\n",
    "                legible_name = configs['legible_name'][subtype]\n",
    "                if legible_name in to_include:\n",
    "                    #get metadata for virus subtype\n",
    "                    color = configs['color'][subtype]\n",
    "\n",
    "                    #get rates\n",
    "                    for gene in genes:\n",
    "                        standard_gene = [standard_genes[gene.upper()] if gene.upper() in standard_genes.keys() else False][0]\n",
    "                        gene_len_in_codons = get_rate_in_muts_per_year(virus, subtype, gene, window, min_seqs)\n",
    "                        #get rate with updated outgroup\n",
    "                        #for viruses with under 12 years of data, use smaller windows\n",
    "                        if virus in ['h1n1pdm', 'mumps']:\n",
    "                            window_v, min_seqs_v = 3, 3\n",
    "                            rate_per_codon, lower_95ci, upper_95ci = readin_rates_updated_outgroup(virus_and_subtype, gene, window_v, min_seqs_v)\n",
    "                        else:\n",
    "                            rate_per_codon, lower_95ci, upper_95ci = readin_rates_updated_outgroup(virus_and_subtype, gene, window, min_seqs)\n",
    "                        rate_per_gene = rate_per_codon/1000*gene_len_in_codons\n",
    "                        lower_95ci_per_gene = lower_95ci/1000*gene_len_in_codons\n",
    "                        upper_95ci_per_gene = upper_95ci/1000*gene_len_in_codons\n",
    "\n",
    "                        legible_rate = str(round(rate_per_codon, 2))+'×10⁻³ muts per codon per year'\n",
    "                        legible_rate_per_gene = str(round(rate_per_gene,2))+' muts per year'\n",
    "                        \n",
    "                        #make gene name more legible\n",
    "                        gene_legible = gene.capitalize()\n",
    "                        if virus == 'rotavirus':\n",
    "                            gene_legible = rota_segment_to_gene[gene]\n",
    "                        if gene_legible in legible_gene_names.keys():\n",
    "                            gene_legible = legible_gene_names[gene_legible]\n",
    "                            \n",
    "                        #for polymerase and receptor-binding, make a label with gene name \n",
    "                        if standard_gene == 'polymerase':\n",
    "                            legible_gene_name = f'{gene_legible} (Polymerase)'\n",
    "                        elif standard_gene == 'receptor_binding':\n",
    "                            legible_gene_name = f'{gene_legible} (Receptor-Binding)'\n",
    "                        else:\n",
    "                            legible_gene_name = gene_legible\n",
    "\n",
    "                        all_data.append({'virus':virus, 'subtype':subtype, 'virus_and_subtype':virus_and_subtype, 'gene': gene_legible,\n",
    "                                        'adaptive_subs_per_codon_per_year': rate_per_codon, \n",
    "                                         'lower_95ci': lower_95ci, 'upper_95ci': upper_95ci, 'ci': [lower_95ci, upper_95ci], \n",
    "                                         'len_in_codons': gene_len_in_codons, 'adaptive_muts_per_year': rate_per_gene, \n",
    "                                         'lower_95ci_mutspergene':lower_95ci_per_gene, 'upper_95ci_mutspergene':upper_95ci_per_gene, \n",
    "                                         'legible_name': legible_name, 'virus_family': virus_family.capitalize(), 'color':color, \n",
    "                                         'rna_dna': rna_dna, 'genome_type': genome_type, 'enveloped': enveloped, 'transmission':transmission,\n",
    "                                         'cellular_receptor': cellular_receptor, 'fusion_protein_class': fusion_protein_class, \n",
    "                                         'legible_rate_per_codon': legible_rate, 'legible_rate_per_gene': legible_rate_per_gene, \n",
    "                                         'standard_gene':standard_gene, 'legible_gene_name':legible_gene_name})\n",
    "                        \n",
    "        #if no subtype               \n",
    "        else:\n",
    "            subtype=None\n",
    "            virus_and_subtype = f'{virus}_{subtype}'\n",
    "            #get metadata for virus \n",
    "            color = configs['color']\n",
    "            legible_name = configs['legible_name']\n",
    "            \n",
    "\n",
    "            #get rates\n",
    "            for gene in genes:\n",
    "                standard_gene = [standard_genes[gene.upper()] if gene.upper() in standard_genes.keys() else False][0]\n",
    "                gene_len_in_codons = get_rate_in_muts_per_year(virus, subtype, gene, window, min_seqs)\n",
    "               #get rate with updated outgroup\n",
    "                #for viruses with under 12 years of data, use smaller windows\n",
    "                if virus in ['h1n1pdm', 'mumps']:\n",
    "                    window_v, min_seqs_v = 3, 3\n",
    "                    rate_per_codon, lower_95ci, upper_95ci = readin_rates_updated_outgroup(virus, gene, window_v, min_seqs_v)\n",
    "                else:\n",
    "                    rate_per_codon, lower_95ci, upper_95ci = readin_rates_updated_outgroup(virus, gene, window, min_seqs)\n",
    "                rate_per_gene = rate_per_codon/1000*gene_len_in_codons\n",
    "                lower_95ci_per_gene = lower_95ci/1000*gene_len_in_codons\n",
    "                upper_95ci_per_gene = upper_95ci/1000*gene_len_in_codons\n",
    "                \n",
    "                legible_rate = str(round(rate_per_codon, 2))+'×10⁻³ muts per codon per year'\n",
    "                legible_rate_per_gene = str(round(rate_per_gene,2))+' muts per year'\n",
    "                \n",
    "                #make gene name more legible\n",
    "                gene_legible = gene.capitalize()\n",
    "                if virus == 'rotavirus':\n",
    "                    gene_legible = rota_segment_to_gene[gene]\n",
    "                if gene_legible in legible_gene_names.keys():\n",
    "                    gene_legible = legible_gene_names[gene_legible]\n",
    "                    \n",
    "                #for polymerase and receptor-binding, make a label with gene name \n",
    "                if standard_gene == 'polymerase':\n",
    "                    legible_gene_name = f'{gene_legible} (Polymerase)'\n",
    "                elif standard_gene == 'receptor_binding':\n",
    "                    legible_gene_name = f'{gene_legible} (Receptor-Binding)'\n",
    "                else:\n",
    "                    legible_gene_name = gene_legible\n",
    "            \n",
    "                all_data.append({'virus':virus, 'subtype':subtype, 'virus_and_subtype':virus, 'gene': gene_legible,\n",
    "                                 'adaptive_subs_per_codon_per_year': rate_per_codon, \n",
    "                                 'lower_95ci': lower_95ci, 'upper_95ci': upper_95ci, 'ci': [lower_95ci, upper_95ci], \n",
    "                                 'len_in_codons': gene_len_in_codons, 'adaptive_muts_per_year': rate_per_gene, \n",
    "                                 'lower_95ci_mutspergene':lower_95ci_per_gene, 'upper_95ci_mutspergene':upper_95ci_per_gene, \n",
    "                                 'legible_name': legible_name, 'virus_family': virus_family.capitalize(), 'color':color, \n",
    "                                 'rna_dna': rna_dna, 'genome_type': genome_type, 'enveloped': enveloped, 'transmission':transmission,\n",
    "                                 'cellular_receptor': cellular_receptor, 'fusion_protein_class': fusion_protein_class, \n",
    "                                 'legible_rate_per_codon': legible_rate, 'legible_rate_per_gene': legible_rate_per_gene,\n",
    "                                 'standard_gene':standard_gene, 'legible_gene_name':legible_gene_name})\n",
    "\n",
    "\n",
    "                \n",
    "        \n",
    "    df = pd.DataFrame(all_data)\n",
    "\n",
    "    return df     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b140323b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gather_all_data_constant_outgroup(viruses, to_include, window=5, min_seqs=3):\n",
    "    \"\"\"\n",
    "    Gather all the data for each virus and gene (with rates calculated using the constant outgroup method) \n",
    "    that will be needed for the website\n",
    "    \"\"\"\n",
    "    all_data = []\n",
    "    \n",
    "    #for each virus, find whether there are any subtypes, then get all rates and metadata for these subtypes\n",
    "    for virus in viruses:\n",
    "        configs = readin_virus_config(virus)\n",
    "            \n",
    "        #metadata for virus\n",
    "        genes = configs[\"genes\"]\n",
    "        genome_type = configs['genome_type']\n",
    "        rna_dna = [x for x in ['RNA', 'DNA'] if x in configs['genome_type']][0]\n",
    "        enveloped = configs['enveloped']\n",
    "        transmission = configs['transmission']\n",
    "        cellular_receptor = configs['cellular_receptor']\n",
    "        fusion_protein_class = configs['fusion_protein_class']\n",
    "        virus_family = configs['virus_family']\n",
    "        #find the names of the polymerase and receptor-binding genes\n",
    "        polymerase = configs['polymerase']['virus_gene']\n",
    "        receptor_binding = configs['receptor_binding']['virus_gene']\n",
    "        standard_genes = {polymerase.upper():'polymerase', receptor_binding.upper():'receptor_binding'}\n",
    "        \n",
    "        #if virus has subtypes, want to get info for each subtype\n",
    "        if configs['subtype']=='True':\n",
    "            for subtype in configs['subtypes']:\n",
    "                virus_and_subtype = f'{virus}_{subtype}'\n",
    "                legible_name = configs['legible_name'][subtype]\n",
    "                if legible_name in to_include:\n",
    "                    #get metadata for virus subtype\n",
    "                    color = configs['color'][subtype]\n",
    "\n",
    "                    #get rates\n",
    "                    for gene in genes:\n",
    "                        standard_gene = [standard_genes[gene.upper()] if gene.upper() in standard_genes.keys() else False][0]\n",
    "                        gene_len_in_codons = get_rate_in_muts_per_year(virus, subtype, gene, window, min_seqs)\n",
    "                        #get_rate with constant outgroup\n",
    "                        #for viruses with under 12 years of data, use smaller windows\n",
    "                        if virus in ['h1n1pdm', 'mumps']:\n",
    "                            window_v, min_seqs_v = 3, 2\n",
    "                            rate_per_codon, lower_95ci, upper_95ci = readin_rates_constant_outgroup(virus_and_subtype, gene, window_v, min_seqs_v)\n",
    "                        else:\n",
    "                            rate_per_codon, lower_95ci, upper_95ci = readin_rates_constant_outgroup(virus_and_subtype, gene, window, min_seqs)\n",
    "                        rate_per_gene = rate_per_codon/1000*gene_len_in_codons\n",
    "                        lower_95ci_per_gene = lower_95ci/1000*gene_len_in_codons\n",
    "                        upper_95ci_per_gene = upper_95ci/1000*gene_len_in_codons\n",
    "\n",
    "                        legible_rate = str(round(rate_per_codon, 2))+'×10⁻³ muts per codon per year'\n",
    "                        legible_rate_per_gene = str(round(rate_per_gene,2))+' muts per year'\n",
    "                        \n",
    "                        #make gene name more legible\n",
    "                        gene_legible = gene.capitalize()\n",
    "                        if virus == 'rotavirus':\n",
    "                            gene_legible = rota_segment_to_gene[gene]\n",
    "                        if gene_legible in legible_gene_names.keys():\n",
    "                            gene_legible = legible_gene_names[gene_legible]\n",
    "                        \n",
    "                        #for polymerase and receptor-binding, make a label with gene name \n",
    "                        if standard_gene == 'polymerase':\n",
    "                            legible_gene_name = f'{gene_legible} (Polymerase)'\n",
    "                        elif standard_gene == 'receptor_binding':\n",
    "                            legible_gene_name = f'{gene_legible} (Receptor-Binding)'\n",
    "                        else:\n",
    "                            legible_gene_name = gene_legible\n",
    "                        \n",
    "                        all_data.append({'virus':virus, 'subtype':subtype, 'virus_and_subtype':virus_and_subtype, 'gene': gene_legible,\n",
    "                                        'adaptive_subs_per_codon_per_year': rate_per_codon, \n",
    "                                         'lower_95ci': lower_95ci, 'upper_95ci': upper_95ci, 'ci': [lower_95ci, upper_95ci], \n",
    "                                         'len_in_codons': gene_len_in_codons, 'adaptive_muts_per_year': rate_per_gene, \n",
    "                                         'lower_95ci_mutspergene':lower_95ci_per_gene, 'upper_95ci_mutspergene':upper_95ci_per_gene, \n",
    "                                         'legible_name': legible_name, 'virus_family': virus_family.capitalize(), 'color':color, \n",
    "                                         'rna_dna': rna_dna, 'genome_type': genome_type, 'enveloped': enveloped, 'transmission':transmission,\n",
    "                                         'cellular_receptor': cellular_receptor, 'fusion_protein_class': fusion_protein_class, \n",
    "                                         'legible_rate_per_codon': legible_rate, 'legible_rate_per_gene': legible_rate_per_gene, \n",
    "                                         'standard_gene':standard_gene, 'legible_gene_name':legible_gene_name})\n",
    "                        \n",
    "        #if no subtype               \n",
    "        else:\n",
    "            subtype=None\n",
    "            virus_and_subtype = f'{virus}_{subtype}'\n",
    "            #get metadata for virus \n",
    "            color = configs['color']\n",
    "            legible_name = configs['legible_name']\n",
    "            \n",
    "\n",
    "            #get rates\n",
    "            for gene in genes:\n",
    "                standard_gene = [standard_genes[gene.upper()] if gene.upper() in standard_genes.keys() else False][0]\n",
    "                gene_len_in_codons = get_rate_in_muts_per_year(virus, subtype, gene, window, min_seqs)\n",
    "               #get rate with constant outgroup\n",
    "                #for viruses with under 12 years of data, use smaller windows\n",
    "                if virus in ['h1n1pdm', 'mumps']:\n",
    "                    window_v, min_seqs_v = 3, 2\n",
    "                    rate_per_codon, lower_95ci, upper_95ci = readin_rates_constant_outgroup(virus, gene, window_v, min_seqs_v)\n",
    "                else:\n",
    "                    rate_per_codon, lower_95ci, upper_95ci = readin_rates_constant_outgroup(virus, gene, window, min_seqs)\n",
    "                rate_per_gene = rate_per_codon/1000*gene_len_in_codons\n",
    "                lower_95ci_per_gene = lower_95ci/1000*gene_len_in_codons\n",
    "                upper_95ci_per_gene = upper_95ci/1000*gene_len_in_codons\n",
    "                \n",
    "                legible_rate = str(round(rate_per_codon, 2))+'×10⁻³ muts per codon per year'\n",
    "                legible_rate_per_gene = str(round(rate_per_gene,2))+' muts per year'\n",
    "                \n",
    "                #make gene name more legible\n",
    "                gene_legible = gene.capitalize()\n",
    "                if virus == 'rotavirus':\n",
    "                    gene_legible = rota_segment_to_gene[gene]\n",
    "                if gene_legible in legible_gene_names.keys():\n",
    "                    gene_legible = legible_gene_names[gene_legible]\n",
    "                #for polymerase and receptor-binding, make a label with gene name \n",
    "                if standard_gene == 'polymerase':\n",
    "                    legible_gene_name = f'{gene_legible} (Polymerase)'\n",
    "                elif standard_gene == 'receptor_binding':\n",
    "                    legible_gene_name = f'{gene_legible} (Receptor-Binding)'\n",
    "                else:\n",
    "                    legible_gene_name = gene_legible\n",
    "            \n",
    "                all_data.append({'virus':virus, 'subtype':subtype, 'virus_and_subtype':virus, 'gene': gene_legible,\n",
    "                                 'adaptive_subs_per_codon_per_year': rate_per_codon, \n",
    "                                 'lower_95ci': lower_95ci, 'upper_95ci': upper_95ci, 'ci': [lower_95ci, upper_95ci], \n",
    "                                 'len_in_codons': gene_len_in_codons, 'adaptive_muts_per_year': rate_per_gene, \n",
    "                                 'lower_95ci_mutspergene':lower_95ci_per_gene, 'upper_95ci_mutspergene':upper_95ci_per_gene, \n",
    "                                 'legible_name': legible_name, 'virus_family': virus_family.capitalize(), 'color':color, \n",
    "                                 'rna_dna': rna_dna, 'genome_type': genome_type, 'enveloped': enveloped, 'transmission':transmission, \n",
    "                                 'cellular_receptor': cellular_receptor, 'fusion_protein_class': fusion_protein_class, \n",
    "                                 'legible_rate_per_codon': legible_rate, 'legible_rate_per_gene': legible_rate_per_gene,\n",
    "                                 'standard_gene':standard_gene, 'legible_gene_name':legible_gene_name})\n",
    "\n",
    "\n",
    "                \n",
    "        \n",
    "    df = pd.DataFrame(all_data)\n",
    "\n",
    "    return df     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5a8104e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_update = gather_all_data_update_outgroup(['h3n2', 'h1n1pdm','vic', 'yam', 'influenzaC', \n",
    "                    'measles', 'mumps', 'parainfluenza', 'rsv',\n",
    "                    'oc43', '229e', 'nl63',\n",
    "                    'dengue', 'rotavirusAg1p8', 'rotavirusAg1p4', 'norovirus',\n",
    "                    'enterovirusd68', 'hepatitisA_IA',\n",
    "                    'hepatitisB','parvovirusB19', 'adenovirusB7', 'adenovirusB3'], \n",
    "                to_include = ['Influenza A/H3N2', 'Influenza A/H1N1pdm', 'Influenza B/Vic', 'Influenza B/Yam', \n",
    "                              'Influenza C/Yamagata', 'Measles', 'Mumps', 'Parainfluenza-1', 'Parainfluenza-3', \n",
    "                              'RSV-A', 'RSV-B', \n",
    "                              'OC43-A', '229E', 'NL63', \n",
    "                              'Dengue 1-V', 'Dengue 2-AA', 'Dengue 3-III', 'Dengue 4-II', \n",
    "                               'Rotavirus A/P[8]', 'Rotavirus A/P[4]', 'Norovirus GII.4', 'Enterovirus D68', \n",
    "                              'Hepatitis A-IA', 'Parvovirus B19-1A', \n",
    "                              'Adenovirus B-7', 'Adenovirus B-3', 'Hepatitis B-A2', 'Hepatitis B-D3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "019baaf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_constant = gather_all_data_constant_outgroup(['h3n2', 'h1n1pdm','vic', 'yam', 'influenzaC', \n",
    "                    'measles', 'mumps', 'parainfluenza', 'rsv',\n",
    "                    'oc43', '229e', 'nl63',\n",
    "                    'dengue', 'rotavirusAg1p8', 'rotavirusAg1p4', 'norovirus',\n",
    "                    'enterovirusd68', 'hepatitisA_IA',\n",
    "                    'hepatitisB','parvovirusB19', 'adenovirusB7', 'adenovirusB3'], \n",
    "                to_include = ['Influenza A/H3N2', 'Influenza A/H1N1pdm', 'Influenza B/Vic', 'Influenza B/Yam', \n",
    "                              'Influenza C/Yamagata', 'Measles', 'Mumps', 'Parainfluenza-1', 'Parainfluenza-3', \n",
    "                              'RSV-A', 'RSV-B', \n",
    "                              'OC43-A', '229E', 'NL63', \n",
    "                              'Dengue 1-V', 'Dengue 2-AA', 'Dengue 3-III', 'Dengue 4-II', \n",
    "                               'Rotavirus A/P[8]', 'Rotavirus A/P[4]', 'Norovirus GII.4', 'Enterovirus D68', \n",
    "                              'Hepatitis A-IA', 'Parvovirus B19-1A', \n",
    "                              'Adenovirus B-7', 'Adenovirus B-3', 'Hepatitis B-A2', 'Hepatitis B-D3'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe76167",
   "metadata": {},
   "source": [
    "Now save all the data to .csv files. Want to partition data into specific files that will make plotting in d3 easier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6ebfbc50",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_polyermase_update_outgroup(filename):\n",
    "    \"\"\"\n",
    "    Save a csv file to be read in by d3 that contains all data for the polymerase gene, \n",
    "    with rates calculated using the update outgroup methods\n",
    "    \"\"\"\n",
    "    \n",
    "    #partition df to only polymerase\n",
    "    df_p = df_update[df_update['standard_gene']=='polymerase']\n",
    "    \n",
    "    \n",
    "    df_p.to_csv(filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "452845bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_polyermase_update_outgroup('aggregated_data_for_d3/polymerase_updateog_for_d3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5ef9b0bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_polyermase_constant_outgroup(filename):\n",
    "    \"\"\"\n",
    "    Save a csv file to be read in by d3 that contains all data for the polymerase gene, \n",
    "    with rates calculated using the constant outgroup methods\n",
    "    \"\"\"\n",
    "    \n",
    "    #partition df to only polymerase\n",
    "    df_p = df_constant[df_constant['standard_gene']=='polymerase']\n",
    "    \n",
    "    \n",
    "    df_p.to_csv(filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e30aa845",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_polyermase_constant_outgroup('aggregated_data_for_d3/polymerase_constantog_for_d3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2b4fe81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_rb_update_outgroup(filename):\n",
    "    \"\"\"\n",
    "    Save a csv file to be read in by d3 that contains all data for the receptor-binding gene, \n",
    "    with rates calculated using the update outgroup methods\n",
    "    \"\"\"\n",
    "    \n",
    "    #partition df to only rb\n",
    "    df_rb = df_update[df_update['standard_gene']=='receptor_binding']\n",
    "    \n",
    "    \n",
    "    df_rb.to_csv(filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6c610bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_rb_update_outgroup('aggregated_data_for_d3/rb_updateog_for_d3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1c97d3e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_rb_constant_outgroup(filename):\n",
    "    \"\"\"\n",
    "    Save a csv file to be read in by d3 that contains all data for the receptor-binding gene, \n",
    "    with rates calculated using the constant outgroup methods\n",
    "    \"\"\"\n",
    "    \n",
    "    #partition df to only rb\n",
    "    df_rb = df_constant[df_constant['standard_gene']=='receptor_binding']\n",
    "    \n",
    "    \n",
    "    df_rb.to_csv(filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b1b11ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_rb_constant_outgroup('aggregated_data_for_d3/rb_constantog_for_d3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d347a21f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_all_update_outgroup(filename):\n",
    "    \"\"\"\n",
    "    Save a csv file to be read in by d3 that contains all data for all viruses and all genes,\n",
    "    with rates calculated using the update outgroup method\n",
    "    \"\"\"\n",
    "    \n",
    "    df_update.to_csv(filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9ab4f2cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_all_update_outgroup('aggregated_data_for_d3/all_genes_updateog.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7ccd9336",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_all_constant_outgroup(filename):\n",
    "    \"\"\"\n",
    "    Save a csv file to be read in by d3 that contains all data for all viruses and all genes,\n",
    "    with rates calculated using the constant outgroup method\n",
    "    \"\"\"\n",
    "    \n",
    "    df_constant.to_csv(filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cdd2b060",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_all_constant_outgroup('aggregated_data_for_d3/all_genes_constantog.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffcdc214",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
